{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b2107392",
   "metadata": {},
   "source": [
    "# Analysis of Cohort 1 Samples\n",
    "----------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "## Long-Read Sequencing of CHOP Cancer Fusion Panel Short-Read Libraries - GF Filtering Criteria Pipeline: \n",
    "\n",
    "### Description:\n",
    "This interactive Jupyter Notebook provides the comprehensive analysis pipeline for ONT long-read sequencing of short-read libraries previously processed through the CHOP Cancer Fusion Panel, with a focus on identifying gene fusions (GFs). \n",
    "\n",
    "This workflow incorporates GF detection and filtering criteria to help prioritize biologically and clinically relevant GFs. After each filtering step, there is a count of the number of GFs in each sample remaining. The notebook also incorporates cross-referencing with external databases, and a summary of key findings.\n",
    "\n",
    "\n",
    "### *This notebook can be used for the Analysis and Filtering of Cohort 1 Subset Samples - the Input and Output paths need to updated\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------------------\n",
    "## Table of Contents\n",
    "\n",
    "### 1. Set Inputs\n",
    "- **a.** Import Libraries and Pandas Display Options  \n",
    "- **b.** GENCODE and Genome Reference Files  \n",
    "- **c.** Set Sample Names List \n",
    "\n",
    "\n",
    "### 2. Offline Scripts\n",
    "- **a.** Redo Basecalling\n",
    "- **b.** Alignment to GRCh38 Human Reference Genome\n",
    "- **c.** GF Detection Programs (LongGF, JAFFAL, FusionSeeker)\n",
    "- **d.** Generate GF Summary of Detected GFs (01_Combine_Fusion_Program_Outputs_nb.ipynb)\n",
    "\n",
    "\n",
    "### 3. Gene Fusion Detection Results & Filtering Criteria: Input Summaries of LongGF, JAFFAL, and FusionSeeker Outputs  \n",
    "- **a.** Keep GFs involving CHOP Cancer Fusion Panel and Cancer Census Genes  \n",
    "- **b.** Keep Fusions with More than 2 Reads (10 reads for Cohort 1 Subset Samples)\n",
    "- **c.** Remove Fusions with Unrelated Genes\n",
    "- **d.** Keep Fusions with Consistent Strandness   \n",
    "\n",
    "### 4. Fusion Annotation\n",
    "- **a.** Cross-Referencing Detected Fusions with Online Fusion Databases  \n",
    "- **b.** Remove Recurrent Fusions in More Than a Certain Percentage of the Samples (Excluding those found in Online Fusion Databases) \n",
    "\n",
    "\n",
    "### 5. Results Summary\n",
    "- **a.** Summary of Results  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbeefeee",
   "metadata": {},
   "source": [
    "# 1. Set Inputs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a61f231d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyensembl\n",
    "import pickle\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Pandas Display Options for better readability\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0e92a53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# File paths and parameters for GENCODE Reference and Annotation\n",
    "# Downloaded: https://www.gencodegenes.org/human/release_46.html (using release preference)\n",
    "# basic annotation GTF, transcript fasta, and protein fasta\n",
    "\n",
    "reference_name = 'GRCh38'\n",
    "annotation_name = 'gencode'\n",
    "annotation_version = 'v46' # recommended\n",
    "\n",
    "# File paths for GTF, transcript, and protein fasta files\n",
    "gtf_path = '/path/to/gencode.v46.basic.annotation.gtf.gz'\n",
    "transcript_fasta_path = '/path/to/gencode.v46.transcripts.fa.gz'\n",
    "protein_fasta_path = '/path/to/gencode.v46.pc_transcripts.fa.gz'\n",
    "\n",
    "# Cache directory path for downloading and storing files\n",
    "cache_directory_path = '/path/to/gencode_v46_downloads_cache/'\n",
    "\n",
    "### \n",
    "# GFF3 v46 Annotation File for Reference Lookups\n",
    "gff3_input_annotation = '/path/to/gencode.v46.basic.annotation.gff3'\n",
    "\n",
    "### \n",
    "# Set names of samples included in analysis\n",
    "sample_names = ['sample3']\n",
    "print(\"Sample Names: \\n\", sample_names)\n",
    "\n",
    "###\n",
    "# Set path to your summary of all fusions file\n",
    "input_sum_GF_sample_path = '/path/to/folder/combined_fusion_calls/'\n",
    "\n",
    "###\n",
    "# Set Path to CHOP Cancer Fusion Panel and Cancer Census Gene Lists\n",
    "chop_panel_genes_path = '/path/to/CHOP_Fusion_Panel_genes.txt'\n",
    "cancer_census_genes_path = '/path/to/Cosmic_CancerGeneCensus_v101_GRCh38.tsv'\n",
    "\n",
    "###\n",
    "# Set Read Support Threshold \n",
    "set_read_support = 2 # change based on read support needs\n",
    "\n",
    "###\n",
    "# Set Path to Literature Reference Files \n",
    "mitelman_db_gfs = \"/path/to/mitelman_db_fusion_names.txt\"\n",
    "chimer_db_gfs = \"/path/to/ChimerDB4.0_fusions.txt\"\n",
    "cosmic_db_gfs = \"/path/to/cosmic_fusion_genes_cleaned.tsv\"\n",
    "\n",
    "###\n",
    "# Set Recurrent Percentage to Filter Fusions On\n",
    "percentage_samples = 0.15\n",
    "\n",
    "###\n",
    "# Set Path to Folder to Save Output Summary Files \n",
    "filtered_output_save_folder = '/path/to/filtered_fusion_calls/'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25459fad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# File paths and parameters for GENCODE Reference and Annotation\n",
    "# Downloaded: https://www.gencodegenes.org/human/release_46.html (using release preference)\n",
    "# basic annotation GTF, transcript fasta, and protein fasta\n",
    "\n",
    "reference_name = 'GRCh38'\n",
    "annotation_name = 'gencode'\n",
    "annotation_version = 'v46' # recommended\n",
    "\n",
    "# File paths for GTF, transcript, and protein fasta files\n",
    "gtf_path = '/path/to/gencode.v46.basic.annotation.gtf.gz'\n",
    "transcript_fasta_path = '/path/to/gencode.v46.transcripts.fa.gz'\n",
    "protein_fasta_path = '/path/to/gencode.v46.pc_transcripts.fa.gz'\n",
    "\n",
    "# Cache directory path for downloading and storing files\n",
    "cache_directory_path = '/path/to/gencode_v46_downloads_cache/'\n",
    "\n",
    "\n",
    "########\n",
    "reference_name = 'GRCh38'\n",
    "annotation_name = 'gencode'\n",
    "annotation_version = 'v46'\n",
    "\n",
    "# File paths for GTF, transcript, and protein fasta files\n",
    "gtf_path = '/mnt/isilon/wang_lab/karly/ref_genome_hg38/cleaned_hg38_ref/gencode.v46.basic.annotation.gtf.gz'\n",
    "transcript_fasta_path = '/mnt/isilon/wang_lab/karly/ref_genome_hg38/cleaned_hg38_ref/gencode.v46.transcripts.fa.gz'\n",
    "protein_fasta_path = '/mnt/isilon/wang_lab/karly/ref_genome_hg38/cleaned_hg38_ref/gencode.v46.pc_transcripts.fa.gz'\n",
    "\n",
    "# Cache directory path for downloading and storing files\n",
    "cache_directory_path = '/mnt/isilon/wang_lab/karly/ref_genome_hg38/cleaned_hg38_ref/gencode_v46_downloads_cache/'\n",
    "########\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### \n",
    "\n",
    "# GFF3 v46 Annotation File for Reference Lookups\n",
    "gff3_input_annotation = '/path/to/gencode.v46.basic.annotation.gff3'\n",
    "gff3_input_annotation = '/mnt/isilon/wang_lab/karly/ref_genome_hg38/cleaned_hg38_ref/gencode.v46.basic.annotation.gff3'\n",
    "\n",
    "\n",
    "### \n",
    "\n",
    "# Set names of samples included in analysis\n",
    "sample_names = ['sample3']\n",
    "print(\"Sample Names: \\n\", sample_names)\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "# Define the path to your summary of all fusions file\n",
    "input_sum_GF_sample_path = '/mnt/isilon/wang_lab/karly/project/chop_panel_ont_aim1/github_upload/example_chop_fusion_analysis/combined_fusion_calls/'\n",
    "\n",
    "###\n",
    "chop_panel_genes_path = '/path/to/CHOP_Fusion_Panel_genes.txt'\n",
    "cancer_census_genes_path = '/path/to/Cosmic_CancerGeneCensus_v101_GRCh38.tsv'\n",
    "\n",
    "\n",
    "chop_panel_genes_path = '/mnt/isilon/wang_lab/karly/project/chop_panel_ont_aim1/github_upload/references/CHOP_Fusion_Panel_genes.txt'\n",
    "cancer_census_genes_path = '/mnt/isilon/wang_lab/karly/project/chop_panel_ont_aim1/github_upload/references/Cosmic_CancerGeneCensus_v101_GRCh38.tsv'\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "set_read_support = 2 # change based on read support needs\n",
    "\n",
    "###\n",
    "\n",
    "# Literature Cross Reference \n",
    "mitelman_db_gfs = \"/path/to/mitelman_db_fusion_names.txt\"\n",
    "chimer_db_gfs = \"/path/to/ChimerDB4.0_fusions.txt\"\n",
    "cosmic_db_gfs = \"/path/to/cosmic_fusion_genes_cleaned.tsv\"\n",
    "\n",
    "\n",
    "mitelman_db_gfs = '/mnt/isilon/wang_lab/karly/project/chop_panel_ont_aim1/github_upload/references/mitelman_fusion_names.txt'\n",
    "chimer_db_gfs = '/mnt/isilon/wang_lab/karly/project/chop_panel_ont_aim1/github_upload/references/ChimerDB4.0_fusions.txt'\n",
    "cosmic_db_gfs = '/mnt/isilon/wang_lab/karly/project/chop_panel_ont_aim1/github_upload/references/cosmic_fusion_genes_cleaned.tsv'\n",
    "\n",
    "###\n",
    "\n",
    "# set recurrent percentage to filter fusions from \n",
    "percentage_samples = 0.15\n",
    "\n",
    "###\n",
    "\n",
    "filtered_output_save_folder = '/mnt/isilon/wang_lab/karly/project/chop_panel_ont_aim1/github_upload/example_chop_fusion_analysis/filtered_fusion_calls/'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93adf6e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize the pyensembl Genome object with the defined variables\n",
    "gencode_data = pyensembl.Genome(\n",
    "    reference_name=reference_name,\n",
    "    annotation_name=annotation_name,\n",
    "    annotation_version=annotation_version,\n",
    "    gtf_path_or_url=gtf_path,\n",
    "    transcript_fasta_paths_or_urls=transcript_fasta_path,\n",
    "    protein_fasta_paths_or_urls=protein_fasta_path,\n",
    "    cache_directory_path=cache_directory_path,\n",
    "    copy_local_files_to_cache=False)\n",
    "\n",
    "\n",
    "# Index the genome data to enable quick access to annotations\n",
    "print('gencode reference stored')\n",
    "gencode_data.index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c834a98c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Process GFF3 v46 annotation file for reference lookups, extracting key details.\n",
    "# Load the GFF3 file into a DataFrame, skipping commented lines and using tab-delimited columns.\n",
    "annotation_ref_df = pd.read_csv(gff3_input_annotation, sep='\\t', comment='#', header=None)\n",
    "\n",
    "# Rename relevant columns for easier access.\n",
    "annotation_ref_df.rename(columns={0: 'chrom', 2: 'feature', 3: 'start', 4: 'end', 6: 'strand', 8: 'info'}, inplace=True)\n",
    "\n",
    "# Extract necessary attributes from the 'info' column using regex.\n",
    "annotation_ref_df['gene_id'] = annotation_ref_df['info'].str.extract(r'gene_id=([^;]+)')\n",
    "annotation_ref_df['gene_name'] = annotation_ref_df['info'].str.extract(r'gene_name=([^;]+)')\n",
    "annotation_ref_df['gene_type'] = annotation_ref_df['info'].str.extract(r'gene_type=([^;]+)')\n",
    "annotation_ref_df['transcript_id'] = annotation_ref_df['info'].str.extract(r'transcript_id=([^;]+)')\n",
    "annotation_ref_df['transcript_name'] = annotation_ref_df['info'].str.extract(r'transcript_name=([^;]+)')\n",
    "annotation_ref_df['transcript_type'] = annotation_ref_df['info'].str.extract(r'transcript_type=([^;]+)')\n",
    "\n",
    "# Create a mapping of (gene_name, chromosome) to all extracted attributes\n",
    "gene_name_map = {\n",
    "    (gene, chrom): (gene_id, strand, gene_type, transcript_id, transcript_name, transcript_type)\n",
    "    for gene, chrom, gene_id, strand, gene_type, transcript_id, transcript_name, transcript_type in zip(\n",
    "        annotation_ref_df['gene_name'], annotation_ref_df['chrom'], annotation_ref_df['gene_id'], annotation_ref_df['strand'], annotation_ref_df['gene_type'],\n",
    "        annotation_ref_df['transcript_id'], annotation_ref_df['transcript_name'], annotation_ref_df['transcript_type']\n",
    "    )\n",
    "}\n",
    "\n",
    "# Function to retrieve gene-related attributes\n",
    "def get_gene_name_map(gene_name, contig):\n",
    "    return gene_name_map.get((gene_name, contig), (None, None, None, None, None, None))\n",
    "\n",
    "# Confirm that the GFF3 file has been processed and the DataFrame is ready for use.\n",
    "print('GFF3 file is stored!')\n",
    "\n",
    "# Display the DataFrame for verification.\n",
    "annotation_ref_df.head()\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61a432f8",
   "metadata": {},
   "source": [
    "# 2. Offline Scripts\n",
    "- **a.** Redo Basecalling\n",
    "- **b.** Alignment to GRCh38 Human Reference Genome\n",
    "- **c.** LongReadSum Alignment Statistics \n",
    "- **d.** GF Detection Programs (LongGF, JAFFAL, FusionSeeker)\n",
    "- **e.** Generate GF Summary of Detected GFs (Combine Fusion Outputs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d141d36",
   "metadata": {},
   "source": [
    "# 3. Gene Fusion Detection Results & Filtering Criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e51460de",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Initialize an empty list to hold the dataframes for each sample\n",
    "combined_df_list = []\n",
    "\n",
    "# Loop through each sample name and process the corresponding files\n",
    "for sample_number in sample_names:\n",
    "    LJF_filter_input_file = f'{input_sum_GF_sample_path}/summary_all_fusions.tsv'\n",
    "    LJF_df = pd.read_csv(LJF_filter_input_file, sep='\\t')\n",
    "    LJF_df['Sample Name'] = sample_number\n",
    "    combined_df_list.append(LJF_df)\n",
    "all_fusions_input_df = pd.concat(combined_df_list, ignore_index=True)\n",
    "\n",
    "\n",
    "# Drop duplicates while keeping the row with the highest 'Supporting Reads'\n",
    "all_fusions_input_df = (\n",
    "    all_fusions_input_df\n",
    "    .sort_values('Supporting Reads', ascending=False)  # Sort by 'Supporting Reads' in descending order\n",
    "    .drop_duplicates(subset=['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol', \n",
    "                              'Gene 1 Breakpoint', 'Gene 2 Breakpoint', \n",
    "                              'GF Program', 'Sample Name'], \n",
    "                     keep='first'))  # Retain the first occurrence (highest 'Supporting Reads')\n",
    "\n",
    "print(\"Total Fusions Detected: \", all_fusions_input_df.shape[0])\n",
    "\n",
    "# number of unique fusions \n",
    "unique_fusions_initial_input = all_fusions_input_df[['Fusion Name', 'Sample Name']].drop_duplicates()\n",
    "num_unique_fusions_initial_input = len(unique_fusions_initial_input)\n",
    "print('Unique Gene Fusions Initiall Detected (based on gene name): ', num_unique_fusions_initial_input)\n",
    "\n",
    "all_fusions_input_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb254f40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a df to keep track of fusion counts while filtering\n",
    "fusion_counts_while_filtering = unique_fusions_initial_input.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "fusion_counts_while_filtering.columns = ['Sample Name', 'Initial Detected Unique GFs']\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96233e13",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d8caf5e7",
   "metadata": {},
   "source": [
    "## **a.** Keep GFs with CHOP Cancer Panel Genes and Cancer Census Union of Genes  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46dae952",
   "metadata": {},
   "outputs": [],
   "source": [
    "chop_genes_df = pd.read_csv(chop_panel_genes_path, header=None)\n",
    "chop_genes = set(chop_genes_df[0].dropna())  \n",
    "print(f\"Total number of CHOP genes: {len(chop_genes)}\")\n",
    "\n",
    "###\n",
    "\n",
    "cancer_census_df = pd.read_csv(cancer_census_genes_path, sep='\\t', usecols=['GENE_SYMBOL'])\n",
    "cc_genes = set(cancer_census_df['GENE_SYMBOL'].dropna())\n",
    "print('Number of Cancer Census Genes: ', len(cc_genes))\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "# Combine Genes\n",
    "combined_genes_set = chop_genes.union(cc_genes)\n",
    "combined_genes_list = set(combined_genes_set)\n",
    "\n",
    "# Look at overlap\n",
    "overlap_with_chop = chop_genes.intersection(cc_genes)\n",
    "print(f\"Total number of combined genes: {len(combined_genes_list)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a61d7e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter fusions to those that involve at least one gene in the union of CHOP & Cancer Census genes\n",
    "chop_cancer_census_gfs = all_fusions_input_df[\n",
    "    all_fusions_input_df['Gene 1 Symbol'].isin(combined_genes_list) | \n",
    "    all_fusions_input_df['Gene 2 Symbol'].isin(combined_genes_list)]\n",
    "\n",
    "# Note which gene is involved\n",
    "def check_overlap(genes_set, column_name):\n",
    "    def check(row):\n",
    "        gene1_in_panel = row['Gene 1 Symbol'] in genes_set\n",
    "        gene2_in_panel = row['Gene 2 Symbol'] in genes_set\n",
    "        if gene1_in_panel and gene2_in_panel:\n",
    "            return 'Both'\n",
    "        elif gene1_in_panel:\n",
    "            return 'Gene 1'\n",
    "        elif gene2_in_panel:\n",
    "            return 'Gene 2'\n",
    "        else:\n",
    "            return 'None'\n",
    "    chop_cancer_census_gfs_copy = chop_cancer_census_gfs.copy()\n",
    "    chop_cancer_census_gfs_copy[column_name] = chop_cancer_census_gfs_copy.apply(check, axis=1)\n",
    "\n",
    "check_overlap(chop_genes, 'CHOP Fusion Panel Overlap')\n",
    "check_overlap(cc_genes, 'Cancer Census Overlap')\n",
    "\n",
    "print(f\"Number of fusions with at least 1 CHOP or Cancer Census gene: {chop_cancer_census_gfs.shape[0]}\")\n",
    "chop_cancer_census_gfs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d241891",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_chop_filter1 = chop_cancer_census_gfs.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_chop_filter1, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Filter on CHOP and Cancer Census Genes'}, inplace=True)\n",
    "fusion_counts_while_filtering['Filter on CHOP and Cancer Census Genes'] = fusion_counts_while_filtering['Filter on CHOP and Cancer Census Genes'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "def00f2c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "84b53021",
   "metadata": {},
   "source": [
    "## **b.** Keep Fusions with More than 2 Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eff70534",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove those with less than 2 supporting reads\n",
    "read_filter_df = chop_cancer_census_gfs[chop_cancer_census_gfs['Supporting Reads'] >= set_read_support]\n",
    "\n",
    "# Number of entries removed\n",
    "num_removed_read_filter = chop_cancer_census_gfs.shape[0] - read_filter_df.shape[0]\n",
    "\n",
    "print(f'Number of fusions with {set_read_support} or more supporting reads: {read_filter_df.shape[0]}')\n",
    "print(\"Number of fusions removed:\", num_removed_read_filter)\n",
    "read_filter_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aedf3889",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_read_filter2 = read_filter_df.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_read_filter2, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Filter on Read Support'}, inplace=True)\n",
    "fusion_counts_while_filtering['Filter on Read Support'] = fusion_counts_while_filtering['Filter on Read Support'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f98771",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d0b4e5d4",
   "metadata": {},
   "source": [
    "## **c.** Remove Fusions with Unrelated Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d14721a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove Pseudogenes\n",
    "\n",
    "# Filter rows where 'feature' column contains the word 'gene'\n",
    "gene_annotation_ref_df = annotation_ref_df[annotation_ref_df['feature'] == 'gene']\n",
    "\n",
    "# Merge Gene 1 gene type\n",
    "def merge_gene_type(df, gene_col, gene_annotation_ref_df, suffix):\n",
    "    merged_df = df.merge(gene_annotation_ref_df[['gene_id', 'gene_type']], \n",
    "                         left_on=gene_col, \n",
    "                         right_on='gene_id', \n",
    "                         how='left')\n",
    "    merged_df[f'{suffix} Gene Type'] = merged_df['gene_type'].fillna('Unknown')\n",
    "    merged_df.drop(columns=['gene_type', 'gene_id'], inplace=True)\n",
    "    return merged_df\n",
    "\n",
    "# Merge Gene 1 data\n",
    "gene1_merged = merge_gene_type(read_filter_df, 'Gene 1 ID', gene_annotation_ref_df, 'Gene 1')\n",
    "\n",
    "# Merge Gene 2 data\n",
    "filtered_gene_types_added = merge_gene_type(gene1_merged, 'Gene 2 ID', gene_annotation_ref_df, 'Gene 2')\n",
    "\n",
    "# Filter for protein-coding genes and exclude pseudogenes\n",
    "filtered_gene_types_protein_coding = filtered_gene_types_added[\n",
    "    (filtered_gene_types_added['Gene 1 Gene Type'] == 'protein_coding') | \n",
    "    (filtered_gene_types_added['Gene 2 Gene Type'] == 'protein_coding')]\n",
    "\n",
    "filtered_gene_types_protein_coding = filtered_gene_types_protein_coding[\n",
    "    ~filtered_gene_types_protein_coding['Gene 1 Gene Type'].str.contains('pseudogene', na=False) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 2 Gene Type'].str.contains('pseudogene', na=False)]\n",
    "\n",
    "\n",
    "# Additionally, remove GFs: # where either 'Gene 1 Breakpoint' or 'Gene 2 Breakpoint' contains 'chrM'\n",
    "# or 'Gene 1 Symbol' or 'Gene 2 Symbol' starts with 'RPL', 'MRP', 'RPS', or 'HLA'\n",
    "\n",
    "# Patterns to filter out\n",
    "chromosome_pattern = ['chrM']\n",
    "ribo_pattern = r'^(RP|RPL|MRP|RPS)'  # Pattern for ribosomal proteins\n",
    "hla_pattern = r'^(HLA)'  # Pattern for HLA genes\n",
    "\n",
    "# Perform the combined filtering to remove unwanted GFs\n",
    "filtered_no_unrelated = filtered_gene_types_protein_coding[\n",
    "    ~filtered_gene_types_protein_coding['Gene 1 Breakpoint'].str.contains('|'.join(chromosome_pattern), na=False) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 2 Breakpoint'].str.contains('|'.join(chromosome_pattern), na=False) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 1 Symbol'].str.match(ribo_pattern) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 2 Symbol'].str.match(ribo_pattern) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 1 Symbol'].str.match(hla_pattern) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 2 Symbol'].str.match(hla_pattern)\n",
    "]\n",
    "\n",
    "# Display the filtered results\n",
    "print(\"Number of fusions remaining:\", filtered_no_unrelated.shape[0])\n",
    "print(\"Number of fusions removed:\", read_filter_df.shape[0] - filtered_no_unrelated.shape[0])\n",
    "filtered_no_unrelated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bef37f80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_gene_type_filter3 = filtered_no_unrelated.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_gene_type_filter3, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Filter on Unrelated Genes'}, inplace=True)\n",
    "fusion_counts_while_filtering['Filter on Unrelated Genes'] = fusion_counts_while_filtering['Filter on Unrelated Genes'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f22036c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "895bb166",
   "metadata": {},
   "source": [
    "##  **d.** Keep Fusions with Consistent Strandness   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dccc1132",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove those with inconsistent strandness \n",
    "consistent_filter_df = filtered_no_unrelated[filtered_no_unrelated['Strand Comparison'] == 'Consistent']\n",
    "\n",
    "# Calculate how many entries were removed\n",
    "num_removed_consistent_filter = filtered_no_unrelated.shape[0] - consistent_filter_df.shape[0]\n",
    "\n",
    "# Display the results\n",
    "print('Number of fusions with Consistent Strandness:', consistent_filter_df.shape[0])\n",
    "print(\"Number of fusions removed:\", num_removed_consistent_filter)\n",
    "consistent_filter_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8a52730",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_consistent_filter4 = consistent_filter_df.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_consistent_filter4, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Filter on Consistent Strandness'}, inplace=True)\n",
    "fusion_counts_while_filtering['Filter on Consistent Strandness'] = fusion_counts_while_filtering['Filter on Consistent Strandness'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ade904cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "646c7872",
   "metadata": {},
   "source": [
    "# 4. Fusion Annotation\n",
    "## **a.** Cross-Referencing Detected Fusions with External Databases \n",
    "- Mitelman, COSMIC, and ChimerDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17d02125",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mitelman DB\n",
    "mitelman_ref = pd.read_csv(mitelman_db_gfs, delimiter='\\t', header=None)\n",
    "mitelman_ref.columns = ['Fusion Name']\n",
    "mitelman_ref['Fusion Name'] = mitelman_ref['Fusion Name'].str.strip()\n",
    "\n",
    "split_fusions = mitelman_ref['Fusion Name'].str.split(':', n=1)  \n",
    "mitelman_ref[['Gene 1', 'Gene 2']] = pd.DataFrame(split_fusions.tolist(), index=mitelman_ref.index)\n",
    "mitelman_ref = mitelman_ref.drop(columns=['Fusion Name'])\n",
    "mitelman_ref.drop_duplicates()\n",
    "\n",
    "\n",
    "# Chimer DB\n",
    "chimer_ref = pd.read_csv(chimer_db_gfs, delimiter='\\t', header=None)\n",
    "chimer_ref.columns = ['Fusion Name']\n",
    "chimer_ref['Fusion Name'] = chimer_ref['Fusion Name'].str.strip()\n",
    "\n",
    "split_fusions = chimer_ref['Fusion Name'].str.split(':', n=1)  \n",
    "chimer_ref[['Gene 1', 'Gene 2']] = pd.DataFrame(split_fusions.tolist(), index=chimer_ref.index)\n",
    "chimer_ref = chimer_ref.drop(columns=['Fusion Name'])\n",
    "chimer_ref.drop_duplicates()\n",
    "\n",
    "\n",
    "# Cosmic DB\n",
    "cosmic_ref = pd.read_csv(cosmic_db_gfs, delimiter='\\t', header=None)\n",
    "cosmic_ref.columns = ['Fusion Name']\n",
    "cosmic_ref['Fusion Name'] = cosmic_ref['Fusion Name'].str.strip()\n",
    "\n",
    "split_fusions = cosmic_ref['Fusion Name'].str.split(':', n=1)  \n",
    "cosmic_ref[['Gene 1', 'Gene 2']] = pd.DataFrame(split_fusions.tolist(), index=cosmic_ref.index)\n",
    "cosmic_ref = cosmic_ref.drop(columns=['Fusion Name'])\n",
    "cosmic_ref.drop_duplicates()\n",
    "\n",
    "\n",
    "# Combine All Fusion Online Database References\n",
    "all_fusions = pd.concat([mitelman_ref, chimer_ref, cosmic_ref])\n",
    "unique_fusions = all_fusions.drop_duplicates(subset=['Gene 1', 'Gene 2'])\n",
    "\n",
    "\n",
    "# Convert the reference files into sets for faster lookup\n",
    "mitelman_set = set(mitelman_ref.apply(lambda row: f\"{row['Gene 1']}:{row['Gene 2']}\", axis=1).dropna())\n",
    "chimer_set = set(chimer_ref.apply(lambda row: f\"{row['Gene 1']}:{row['Gene 2']}\", axis=1).dropna())\n",
    "cosmic_set = set(cosmic_ref.apply(lambda row: f\"{row['Gene 1']}:{row['Gene 2']}\", axis=1).dropna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fd448c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a function to check if fusion is present in any reference file\n",
    "def check_fusion_in_references(row):\n",
    "    fusion1 = f\"{row['Gene 1 Symbol']}:{row['Gene 2 Symbol']}\"\n",
    "    fusion2 = f\"{row['Gene 2 Symbol']}:{row['Gene 1 Symbol']}\"  # Reverse orientation\n",
    "    result = []\n",
    "\n",
    "    # Check against each reference set\n",
    "    if fusion1 in mitelman_set or fusion2 in mitelman_set:\n",
    "        result.append('Mitelman Reference')\n",
    "    if fusion1 in chimer_set or fusion2 in chimer_set:\n",
    "        result.append('Chimer Reference')\n",
    "    if fusion1 in cosmic_set or fusion2 in cosmic_set:\n",
    "        result.append('COSMIC Reference')\n",
    "\n",
    "    # Return the result as a comma-separated string\n",
    "    return ', '.join(result) if result else 'No match'\n",
    "\n",
    "# Apply the function to each row of the dataframe\n",
    "filter_df_literature = consistent_filter_df.copy()\n",
    "filter_df_literature['Literature Cross Reference'] = filter_df_literature.apply(check_fusion_in_references, axis=1)\n",
    "known_gfs_detected_df = filter_df_literature[filter_df_literature['Literature Cross Reference'] != 'No match']\n",
    "\n",
    "\n",
    "# Show overview of known fusions by sample \n",
    "unique_gfs_per_sample = (\n",
    "    known_gfs_detected_df.groupby('Sample Name')['Fusion Name']\n",
    "    .apply(lambda x: \", \".join(sorted(set(x))))  \n",
    "    .reset_index(name='Unique Gene Fusions'))\n",
    "\n",
    "# Add count of unique fusions per sample\n",
    "unique_gfs_per_sample.insert(1, 'Number of Unique Known GFs', unique_gfs_per_sample['Unique Gene Fusions'].apply(lambda x: x.count(\",\") + 1 if x else 0))\n",
    "unique_gfs_per_sample.sort_values(by='Sample Name', inplace=True)\n",
    "\n",
    "\n",
    "# Get the total number of unique fusions\n",
    "num_unique_fusions = len(known_gfs_detected_df['Fusion Name'].value_counts())\n",
    "num_unique_samples = len(known_gfs_detected_df['Sample Name'].value_counts())\n",
    "print(f\"Total unique fusions: {num_unique_fusions} in {num_unique_samples} samples.\")\n",
    "display(unique_gfs_per_sample)\n",
    "\n",
    "\n",
    "# look at those reported in literature only .. just annotation no filtering!\n",
    "display(filter_df_literature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a628355",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_known_ref_filter5 = known_gfs_detected_df.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_known_ref_filter5, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'GFs in Fusion Reference Databases'}, inplace=True)\n",
    "fusion_counts_while_filtering['GFs in Fusion Reference Databases'] = fusion_counts_while_filtering['GFs in Fusion Reference Databases'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "757f8010",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "68fcbdbf",
   "metadata": {},
   "source": [
    "##  **b.** Remove Recurrent Fusions in More Than a Certain Percentage of the Samples  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dad8eda",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print(\"Recurrent Fusions Removed Filter: \")\n",
    "\n",
    "# Count unique samples each fusion appears in\n",
    "fusion_sample_counts = filter_df_literature.groupby(['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol'])['Sample Name'].nunique()\n",
    "\n",
    "# Calculate threshold and filter based on occurrence\n",
    "total_samples = filter_df_literature['Sample Name'].nunique()\n",
    "\n",
    "threshold = total_samples * percentage_samples\n",
    "\n",
    "recurrent_fusions = fusion_sample_counts[fusion_sample_counts > threshold]\n",
    "recurrent_fusions_df = recurrent_fusions.reset_index()\n",
    "recurrent_fusions_df.rename(columns={'Sample Name': 'Sample Count'}, inplace=True)\n",
    "\n",
    "# Output\n",
    "print(f\"Total samples: {total_samples}, using a {percentage_samples*100}% threshold: {threshold} occurrences within a sample\")\n",
    "print(\"Recurrent fusions\")\n",
    "display(recurrent_fusions_df)\n",
    "\n",
    "# Show original entries \n",
    "original_recurrent_entries_df = filter_df_literature[\n",
    "    filter_df_literature[['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol']].apply(\n",
    "        tuple, axis=1\n",
    "    ).isin(recurrent_fusions_df[['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, axis=1))\n",
    "]\n",
    "print(f\"Original entries matching the recurrent fusions: {original_recurrent_entries_df.shape[0]}\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### EXCLUDE: Fusions Identified in Online Databases from Recurrent Removal ###\n",
    "\n",
    "# Identify fusions in literature (direct match or reciprocal)\n",
    "if 'Literature Cross Reference' in filter_df_literature.columns:\n",
    "    literature_fusions = set(\n",
    "        filter_df_literature[\n",
    "            filter_df_literature['Literature Cross Reference'] != 'No match'\n",
    "        ][['Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, axis=1)\n",
    "    )\n",
    "else:\n",
    "    literature_fusions = set()\n",
    "\n",
    "if 'Literature Reciprocal Fusion' in filter_df_literature.columns:\n",
    "    reciprocal_literature_fusions = set(\n",
    "        filter_df_literature[\n",
    "            filter_df_literature['Literature Reciprocal Fusion'] == 'Yes'\n",
    "        ][['Gene 1 Symbol', 'Gene 2 Symbol']].apply(lambda x: (x[1], x[0]), axis=1)\n",
    "    )\n",
    "else:\n",
    "    reciprocal_literature_fusions = set()\n",
    "\n",
    "# Combine both sets\n",
    "literature_fusions |= reciprocal_literature_fusions\n",
    "\n",
    "# Create a set of recurrent fusion pairs (both orientations)\n",
    "recurrent_fusion_pairs = set(\n",
    "    list(zip(recurrent_fusions.index.get_level_values('Gene 1 Symbol'), recurrent_fusions.index.get_level_values('Gene 2 Symbol'))) +\n",
    "    list(zip(recurrent_fusions.index.get_level_values('Gene 2 Symbol'), recurrent_fusions.index.get_level_values('Gene 1 Symbol')))\n",
    ")\n",
    "\n",
    "# Exclude fusions in literature from removal\n",
    "filtered_fusion_pairs = recurrent_fusion_pairs - literature_fusions\n",
    "\n",
    "# Filter out only the **non-literature** recurrent fusions\n",
    "recurrent_filter_df = filter_df_literature[\n",
    "    ~filter_df_literature[['Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, axis=1).isin(filtered_fusion_pairs)\n",
    "]\n",
    "\n",
    "# Results\n",
    "rows_removed = filter_df_literature.shape[0] - recurrent_filter_df.shape[0]\n",
    "print(f\"\\nNumber of rows removed: {rows_removed}\")\n",
    "print(f\"Recurrent GFs Removed, Remaining GFs: {recurrent_filter_df.shape}\")\n",
    "\n",
    "# Show which recurrent fusions were found in literature and not removed\n",
    "literature_recurrent_fusions = recurrent_fusion_pairs & literature_fusions\n",
    "literature_recurrent_fusions_df = filter_df_literature[\n",
    "    filter_df_literature[['Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, axis=1).isin(literature_recurrent_fusions)\n",
    "]\n",
    "\n",
    "print(\"\\nRecurrent fusions found in literature and not removed:\")\n",
    "print(\"\\n\".join(literature_recurrent_fusions_df['Fusion Name'].unique()))\n",
    "\n",
    "recurrent_filter_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c78973eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_recurrent_GF_filter8 = recurrent_filter_df.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_recurrent_GF_filter8, on='Sample Name', how='left')\n",
    "\n",
    "recurrent_col_name = f'Filter on Recurrent in {percentage_samples*100}% of Samples'\n",
    "\n",
    "fusion_counts_while_filtering.rename(\n",
    "    columns={'Fusion Name': recurrent_col_name}, \n",
    "    inplace=True)\n",
    "\n",
    "# Fill NaN values and convert to int\n",
    "fusion_counts_while_filtering[recurrent_col_name] = fusion_counts_while_filtering[recurrent_col_name].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26a11ac2",
   "metadata": {},
   "source": [
    "# 5. Results Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36816f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the Filtered Fusion Entries\n",
    "output_file_path = f'{filtered_output_save_folder}summary_filtered_output.tsv'\n",
    "recurrent_filter_df.to_csv(output_file_path, sep='\\t', index=False)\n",
    "print(f\"File saved to: {output_file_path}\")\n",
    "\n",
    "\n",
    "# Save the Counts of Fusions after each Filtering Step \n",
    "# Specify the output file path\n",
    "output_file_path_counts = f'{filtered_output_save_folder}summary_filtered_output_counts.tsv'\n",
    "fusion_counts_while_filtering.to_csv(output_file_path_counts, sep='\\t', index=False)\n",
    "print(f\"File saved to: {output_file_path_counts}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2eab8eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# look over counts in df: fusion_counts_while_filtering\n",
    "\n",
    "# looking at 'Unique GFs' -- based on gene names - not variable by breakpoint\n",
    "# Drop duplicates based on 'Fusion Name' and 'Sample Name' while keeping the row with the highest 'Supporting Reads'\n",
    "recurrent_filter_df_UNIQUE = recurrent_filter_df.loc[\n",
    "    recurrent_filter_df.groupby(['Fusion Name', 'Sample Name'])['Supporting Reads'].idxmax()]\n",
    "\n",
    "\n",
    "# Total number of GF entries remaining\n",
    "print(\"Results Summary\")\n",
    "print(\"-\" * 50)\n",
    "print(f\"Total number of GF entries remaining: {recurrent_filter_df_UNIQUE.shape[0]}\")\n",
    "print(\"-\" * 50)\n",
    "\n",
    "\n",
    "final_summary_df = fusion_counts_while_filtering\n",
    "\n",
    "# Number of Entries without any GFs remaining:\n",
    "# Define the dynamic column name based on percentage_samples\n",
    "recurrent_col_name = f'Filter on Recurrent in {percentage_samples*100}% of Samples'\n",
    "\n",
    "# Number of Entries without any GFs remaining:\n",
    "no_entries_remaining = final_summary_df[final_summary_df[recurrent_col_name] == 0]\n",
    "print(f\"Number of samples with no fusions: {no_entries_remaining.shape[0]}\", no_entries_remaining['Sample Name'].unique())\n",
    "\n",
    "# Number of Entries with exactly ONE GF remaining:\n",
    "one_entry_remaining = final_summary_df[final_summary_df[recurrent_col_name] == 1]\n",
    "print(f\"Number of samples with only one fusion: {one_entry_remaining.shape[0]}\", one_entry_remaining['Sample Name'].unique())\n",
    "\n",
    "# Number of Entries with MORE THAN ONE GF remaining:\n",
    "more_than_one_entry_remaining = final_summary_df[final_summary_df[recurrent_col_name] > 1]\n",
    "print(f\"Number of samples with more than one fusion: {more_than_one_entry_remaining.shape[0]}\", more_than_one_entry_remaining['Sample Name'].unique())\n",
    "\n",
    "print(\"-\" * 50)\n",
    "\n",
    "\n",
    "\n",
    "# \n",
    "print('Looking over fusion positive samples, below: ')\n",
    "final_summary_df = fusion_counts_while_filtering[~fusion_counts_while_filtering['Sample Name'].isin(['sample21', 'sample25'])]\n",
    "\n",
    "\n",
    "# get min, max, and mean number of fusions across samples\n",
    "min_fusions = int(final_summary_df[recurrent_col_name].min())\n",
    "print(f\"Minimum number of fusions: {min_fusions}\")\n",
    "\n",
    "max_fusions = int(final_summary_df[recurrent_col_name].max())\n",
    "print(f\"Maximum number of fusions: {max_fusions}\")\n",
    "\n",
    "mean_fusions = int(final_summary_df[recurrent_col_name].mean())\n",
    "print(f\"Mean number of fusions: {mean_fusions}\")\n",
    "print(\"-\" * 50)\n",
    "\n",
    "# Count of GF Program entries\n",
    "gf_program_counts = recurrent_filter_df_UNIQUE['GF Program'].value_counts()\n",
    "print(\"GF Program Counts:\")\n",
    "print(gf_program_counts)\n",
    "print(\"-\" * 50)\n",
    "\n",
    "# Average number of supporting reads over remaining fusions:\n",
    "mean_fusions_remaining = int(recurrent_filter_df_UNIQUE['Supporting Reads'].mean())\n",
    "print(f\"Mean number of Reads over Remaining Fusions: {mean_fusions_remaining}\")\n",
    "print(\"-\" * 100)\n",
    "\n",
    "\n",
    "#######################################################################################\n",
    "# Looking over the NON-unique entries -- to see if reported in multiple GF Programs \n",
    "# may have varying breakpoint locations \n",
    "#######################################################################################\n",
    "\n",
    "# Overlap in 3 programs (LongGF, JAFFAL, FusionSeeker)\n",
    "gf_programs = ['LongGF', 'JAFFAL', 'FusionSeeker']\n",
    "filtered_df = recurrent_filter_df[recurrent_filter_df['GF Program'].isin(gf_programs)]\n",
    "\n",
    "# Group by fusion, gene symbols, and sample, and count unique GF Programs\n",
    "program_count = filtered_df.groupby(['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol', 'Sample Name'])['GF Program'].nunique()\n",
    "\n",
    "# Filter only those fusions detected in all three programs\n",
    "three_program_count = program_count[program_count == 3].reset_index()\n",
    "\n",
    "# Merge with original DataFrame to get supporting read count\n",
    "three_program_count = three_program_count.merge(\n",
    "    recurrent_filter_df[['Fusion Name', 'Sample Name', 'Supporting Reads']], \n",
    "    on=['Fusion Name', 'Sample Name'], \n",
    "    how='left')\n",
    "\n",
    "# Remove duplicates caused by merging and get max supporting reads per fusion\n",
    "three_program_count = (\n",
    "    three_program_count.groupby(['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol', 'Sample Name'])\n",
    "    .agg({'GF Program': 'first', 'Supporting Reads': 'max'})  # Take max supporting reads\n",
    "    .reset_index())\n",
    "\n",
    "# Sort by Supporting Reads in descending order\n",
    "three_program_count.sort_values(by='Supporting Reads', ascending=False, inplace=True)\n",
    "\n",
    "# Display results\n",
    "print(f\"Number of fusions reported in all 3 programs (LongGF, JAFFAL, FusionSeeker): {three_program_count.shape[0]}\")\n",
    "print(\"-\" * 100)\n",
    "# display(three_program_count)\n",
    "\n",
    "\n",
    "#######################################################################################\n",
    "\n",
    "# Overlap in only 2 programs (LongGF, JAFFAL, FusionSeeker)\n",
    "print(\"Groups reported in only 2 programs:\")\n",
    "\n",
    "# Filter fusions detected in exactly 2 programs\n",
    "two_program_count = program_count[program_count == 2].reset_index()\n",
    "\n",
    "# Merge to get Supporting Reads from original DataFrame\n",
    "two_program_count = two_program_count.merge(\n",
    "    recurrent_filter_df[['Fusion Name', 'Sample Name', 'Supporting Reads']], \n",
    "    on=['Fusion Name', 'Sample Name'], \n",
    "    how='left')\n",
    "\n",
    "# Remove duplicates and get max Supporting Reads per fusion\n",
    "two_program_count = (\n",
    "    two_program_count.groupby(['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol', 'Sample Name'])\n",
    "    .agg({'GF Program': 'first', 'Supporting Reads': 'max'})  # Keep max supporting reads\n",
    "    .reset_index())\n",
    "\n",
    "# Sort by Supporting Reads in descending order\n",
    "two_program_count.sort_values(by='Supporting Reads', ascending=False, inplace=True)\n",
    "\n",
    "# Print summary and display table\n",
    "print(f\"Number of fusions reported in exactly 2 programs (LongGF, JAFFAL, FusionSeeker): {two_program_count.shape[0]}\")\n",
    "print(\"-\" * 100)\n",
    "# display(two_program_count)\n",
    "\n",
    "#######################################################################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "169ff292",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "fusion_rank",
   "language": "python",
   "name": "fusion_rank"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
