{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b2107392",
   "metadata": {},
   "source": [
    "# Analysis of Cohort 2 Samples\n",
    "\n",
    "## Long-Read Whole-Transcriptome Sequencing of Samples Previously Analyzed by the CHOP Cancer Fusion Panel - GF Filtering Criteria Pipeline: \n",
    "\n",
    "### Description:\n",
    "This interactive Jupyter Notebook provides the comprehensive analysis pipeline for ONT long-read whole transcriptome sequencing of clinical cancer patient samples previously deemed Fusion-Panel Negative by the CHOP Cancer Fusion Panel, with a focus on identifying gene fusions (GFs). \n",
    "\n",
    "This workflow incorporates GF detection and filtering criteria to help prioritize biologically and clinically relevant GFs. After each filtering step, there is a count of the number of GFs in each sample remaining. The notebook also incorporates cross-referencing with external databases, and a summary of the key findings. \n",
    "\n",
    "\n",
    "----------------------------------------------------------------------------------\n",
    "## Table of Contents\n",
    "\n",
    "### 1. Set Inputs\n",
    "- **a.** Import Libraries and Pandas Display Options  \n",
    "- **b.** GENCODE and Genome Reference Files  \n",
    "- **c.** Set Sample Names List \n",
    "\n",
    "\n",
    "### 2. Offline Scripts\n",
    "- **a.** Redo Basecalling\n",
    "- **b.** Alignment to GRCh38 Human Reference Genome\n",
    "- **c.** GF Detection Programs (LongGF, JAFFAL, FusionSeeker)\n",
    "- **d.** Generate GF Summary of Detected GFs (01_Combine_Fusion_Program_Outputs_nb.ipynb)\n",
    "\n",
    "\n",
    "### 3. Gene Fusion Detection Results & Filtering Criteria: Input Summaries of LongGF, JAFFAL, and FusionSeeker Outputs  \n",
    "- **a.** Keep GFs involving CHOP Cancer Fusion Panel and Cancer Census Genes  \n",
    "- **b.** Keep Fusions with More than 4 Reads\n",
    "- **c.** Remove Fusions with Unrelated Genes\n",
    "- **d.** Keep Fusions with Consistent Strandness\n",
    "- **e.** Remove Fusions Detected in Healthy GTEx Brain Tissue Samples\n",
    "\n",
    "\n",
    "### 4. Fusion Annotation\n",
    "- **a.** Cross-Referencing Detected Fusions with Online Fusion Databases  \n",
    "- **b.** Remove Recurrent Fusions in More Than a Certain Percentage of the Samples (Excluding those found in Online Fusion Databases) \n",
    "- **c.** Annotate if GFs involve Genes from GFs associated with CNS and Tumor Morphology Prioritization Using Mitelman DB\n",
    "\n",
    "### 5. Results Summary\n",
    "- **a.** Summary of Results  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbeefeee",
   "metadata": {},
   "source": [
    "# 1. Set Inputs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a61f231d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyensembl\n",
    "import pickle\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Pandas Display Options for better readability\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b995d989",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample Names: \n",
      " ['sampleK562_pool2']\n"
     ]
    }
   ],
   "source": [
    "# File paths and parameters for GENCODE Reference and Annotation\n",
    "# Downloaded: https://www.gencodegenes.org/human/release_46.html (using release preference)\n",
    "# basic annotation GTF, transcript fasta, and protein fasta\n",
    "\n",
    "reference_name = 'GRCh38'\n",
    "annotation_name = 'gencode'\n",
    "annotation_version = 'v46' # recommended\n",
    "\n",
    "# File paths for GTF, transcript, and protein fasta files\n",
    "gtf_path = '/path/to/gencode.v46.basic.annotation.gtf.gz'\n",
    "transcript_fasta_path = '/path/to/gencode.v46.transcripts.fa.gz'\n",
    "protein_fasta_path = '/path/to/gencode.v46.pc_transcripts.fa.gz'\n",
    "\n",
    "# Cache directory path for downloading and storing files\n",
    "cache_directory_path = '/path/to/gencode_v46_downloads_cache/'\n",
    "\n",
    "### \n",
    "# GFF3 v46 Annotation File for Reference Lookups\n",
    "gff3_input_annotation = '/path/to/gencode.v46.basic.annotation.gff3'\n",
    "\n",
    "### \n",
    "# Set names of samples included in analysis\n",
    "sample_names = ['sampleK562_pool2']\n",
    "print(\"Sample Names: \\n\", sample_names)\n",
    "\n",
    "###\n",
    "# Set path to your summary of all fusions file\n",
    "input_sum_GF_sample_path = '/path/to/folder/combined_fusion_calls/'\n",
    "\n",
    "###\n",
    "# Set Path to CHOP Cancer Fusion Panel and Cancer Census Gene Lists\n",
    "chop_panel_genes_path = '/path/to/CHOP_Fusion_Panel_genes.txt'\n",
    "cancer_census_genes_path = '/path/to/Cosmic_CancerGeneCensus_v101_GRCh38.tsv'\n",
    "\n",
    "###\n",
    "# Set Read Support Threshold \n",
    "set_read_support = 4 # change based on read support needs\n",
    "\n",
    "###\n",
    "# Set Path to Fusions Detected in GTEx Brain Tissues\n",
    "gtex_brain_gfs_tsv = '/path/to/unique_GTEx_brain_GFs.tsv'\n",
    "\n",
    "###\n",
    "# Set Path to Genes involved in Fusions Associated with Brain and Brainstem from Mitelman DB\n",
    "mitelman_brain_gfs_csv = \"/path/to/mitelman_db_brain_brainstem_fusions.csv\"  \n",
    "\n",
    "###\n",
    "# Set Path to Literature Reference Files \n",
    "mitelman_db_gfs = \"/path/to/mitelman_db_fusion_names.txt\"\n",
    "chimer_db_gfs = \"/path/to/ChimerDB4.0_fusions.txt\"\n",
    "cosmic_db_gfs = \"/path/to/cosmic_fusion_genes_cleaned.tsv\"\n",
    "\n",
    "###\n",
    "# Set Recurrent Percentage to Filter Fusions On\n",
    "percentage_samples = 0.15\n",
    "\n",
    "###\n",
    "# Set Path to Folder to Save Output Summary Files \n",
    "filtered_output_save_folder = '/path/to/filtered_fusion_calls/'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28a44499",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "25459fad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gencode reference stored\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /mnt/isilon/wang_lab/karly/ref_genome_hg38/cleaned_hg38_ref/gencode_v46_downloads_cache/gencode.v46.transcripts.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /mnt/isilon/wang_lab/karly/ref_genome_hg38/cleaned_hg38_ref/gencode_v46_downloads_cache/gencode.v46.pc_transcripts.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "# Initialize the pyensembl Genome object with the defined variables\n",
    "gencode_data = pyensembl.Genome(\n",
    "    reference_name=reference_name,\n",
    "    annotation_name=annotation_name,\n",
    "    annotation_version=annotation_version,\n",
    "    gtf_path_or_url=gtf_path,\n",
    "    transcript_fasta_paths_or_urls=transcript_fasta_path,\n",
    "    protein_fasta_paths_or_urls=protein_fasta_path,\n",
    "    cache_directory_path=cache_directory_path,\n",
    "    copy_local_files_to_cache=False)\n",
    "\n",
    "\n",
    "# Index the genome data to enable quick access to annotations\n",
    "print('gencode reference stored')\n",
    "gencode_data.index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c834a98c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GFF3 file is stored!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>1</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>5</th>\n",
       "      <th>strand</th>\n",
       "      <th>7</th>\n",
       "      <th>info</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000290825.1;...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L2-202</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L2-202</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L2-202</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L2-202</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom       1     feature  start    end  5 strand  7                                               info            gene_id gene_name gene_type      transcript_id transcript_name transcript_type\n",
       "0  chr1  HAVANA        gene  11869  14409  .      +  .  ID=ENSG00000290825.1;gene_id=ENSG00000290825.1...  ENSG00000290825.1   DDX11L2    lncRNA                NaN             NaN             NaN\n",
       "1  chr1  HAVANA  transcript  11869  14409  .      +  .  ID=ENST00000456328.2;Parent=ENSG00000290825.1;...  ENSG00000290825.1   DDX11L2    lncRNA  ENST00000456328.2     DDX11L2-202          lncRNA\n",
       "2  chr1  HAVANA        exon  11869  12227  .      +  .  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  ENSG00000290825.1   DDX11L2    lncRNA  ENST00000456328.2     DDX11L2-202          lncRNA\n",
       "3  chr1  HAVANA        exon  12613  12721  .      +  .  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  ENSG00000290825.1   DDX11L2    lncRNA  ENST00000456328.2     DDX11L2-202          lncRNA\n",
       "4  chr1  HAVANA        exon  13221  14409  .      +  .  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  ENSG00000290825.1   DDX11L2    lncRNA  ENST00000456328.2     DDX11L2-202          lncRNA"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Process GFF3 v46 annotation file for reference lookups, extracting key details.\n",
    "# Load the GFF3 file into a DataFrame, skipping commented lines and using tab-delimited columns.\n",
    "annotation_ref_df = pd.read_csv(gff3_input_annotation, sep='\\t', comment='#', header=None)\n",
    "\n",
    "# Rename relevant columns for easier access.\n",
    "annotation_ref_df.rename(columns={0: 'chrom', 2: 'feature', 3: 'start', 4: 'end', 6: 'strand', 8: 'info'}, inplace=True)\n",
    "\n",
    "# Extract necessary attributes from the 'info' column using regex.\n",
    "annotation_ref_df['gene_id'] = annotation_ref_df['info'].str.extract(r'gene_id=([^;]+)')\n",
    "annotation_ref_df['gene_name'] = annotation_ref_df['info'].str.extract(r'gene_name=([^;]+)')\n",
    "annotation_ref_df['gene_type'] = annotation_ref_df['info'].str.extract(r'gene_type=([^;]+)')\n",
    "annotation_ref_df['transcript_id'] = annotation_ref_df['info'].str.extract(r'transcript_id=([^;]+)')\n",
    "annotation_ref_df['transcript_name'] = annotation_ref_df['info'].str.extract(r'transcript_name=([^;]+)')\n",
    "annotation_ref_df['transcript_type'] = annotation_ref_df['info'].str.extract(r'transcript_type=([^;]+)')\n",
    "\n",
    "# Create a mapping of (gene_name, chromosome) to all extracted attributes\n",
    "gene_name_map = {\n",
    "    (gene, chrom): (gene_id, strand, gene_type, transcript_id, transcript_name, transcript_type)\n",
    "    for gene, chrom, gene_id, strand, gene_type, transcript_id, transcript_name, transcript_type in zip(\n",
    "        annotation_ref_df['gene_name'], annotation_ref_df['chrom'], annotation_ref_df['gene_id'], annotation_ref_df['strand'], annotation_ref_df['gene_type'],\n",
    "        annotation_ref_df['transcript_id'], annotation_ref_df['transcript_name'], annotation_ref_df['transcript_type']\n",
    "    )\n",
    "}\n",
    "\n",
    "# Function to retrieve gene-related attributes\n",
    "def get_gene_name_map(gene_name, contig):\n",
    "    return gene_name_map.get((gene_name, contig), (None, None, None, None, None, None))\n",
    "\n",
    "# Confirm that the GFF3 file has been processed and the DataFrame is ready for use.\n",
    "print('GFF3 file is stored!')\n",
    "\n",
    "# Display the DataFrame for verification.\n",
    "annotation_ref_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61a432f8",
   "metadata": {},
   "source": [
    "# 2. Offline Scripts\n",
    "- **a.** Redo Basecalling\n",
    "- **b.** Alignment to GRCh38 Human Reference Genome\n",
    "- **c.** GF Detection Programs (LongGF, JAFFAL, FusionSeeker)\n",
    "- **d.** Generate GF Summary of Detected GFs (01_Combine_Fusion_Program_Outputs_nb.ipynb)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d141d36",
   "metadata": {},
   "source": [
    "# 3. Gene Fusion Detection Results & Filtering Criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e51460de",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Fusions Detected:  4915\n",
      "Unique Gene Fusions Initiall Detected (based on gene name):  4031\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 1 ID</th>\n",
       "      <th>Gene 1 Strand</th>\n",
       "      <th>Gene 1 Reference Strand</th>\n",
       "      <th>Gene 1 Breakpoint</th>\n",
       "      <th>Gene 1 Input Position</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Gene 2 ID</th>\n",
       "      <th>Gene 2 Strand</th>\n",
       "      <th>Gene 2 Reference Strand</th>\n",
       "      <th>Gene 2 Breakpoint</th>\n",
       "      <th>Gene 2 Input Position</th>\n",
       "      <th>Breakpoint Distance</th>\n",
       "      <th>Supporting Reads</th>\n",
       "      <th>Read IDs</th>\n",
       "      <th>GF Program</th>\n",
       "      <th>Strand Comparison</th>\n",
       "      <th>Sample Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23130</th>\n",
       "      <td>MT-ND4L:MT-ND4</td>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chrM:10689</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chrM:10727</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.800000e+01</td>\n",
       "      <td>610</td>\n",
       "      <td>ff94284e-2348-4212-b7f0-50432d204860,0116346f-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23124</th>\n",
       "      <td>PSPHP1:KMT2C</td>\n",
       "      <td>PSPHP1</td>\n",
       "      <td>ENSG00000226278.1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7:55773179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>ENSG00000055609.21</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:152367259</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.659408e+07</td>\n",
       "      <td>147</td>\n",
       "      <td>fe0bb8e8-c947-46d5-a13e-79a50b3a59d2,0096cbc0-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23111</th>\n",
       "      <td>BAG6:SLC44A4</td>\n",
       "      <td>BAG6</td>\n",
       "      <td>ENSG00000204463.14</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:31651655</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SLC44A4</td>\n",
       "      <td>ENSG00000204385.13</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:31865786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.141310e+05</td>\n",
       "      <td>114</td>\n",
       "      <td>ff508124-0c29-4076-bc09-5f000bd29808,000d6e61-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12765</th>\n",
       "      <td>BAG6:SLC44A4</td>\n",
       "      <td>BAG6</td>\n",
       "      <td>ENSG00000204463.14</td>\n",
       "      <td>+,-,-,-,-,-,+,-,-,+,+,-,-,+,+,-,-,-,+,+,+,+,-,...</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:31651657</td>\n",
       "      <td>1167-1500,38-383,37-376,28-291,35-366,30-360,1...</td>\n",
       "      <td>SLC44A4</td>\n",
       "      <td>ENSG00000204385.13</td>\n",
       "      <td>+,-,-,-,-,-,+,-,-,+,+,-,-,+,+,-,-,-,+,+,+,+,-,...</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:31865784</td>\n",
       "      <td>80-1156,381-2334,374-1444,292-1360,367-1426,36...</td>\n",
       "      <td>2.141270e+05</td>\n",
       "      <td>93</td>\n",
       "      <td>000d6e61-632c-461e-b373-82d4fa281a8f,08cbd1ba-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21314</th>\n",
       "      <td>BAG6:SLC44A4</td>\n",
       "      <td>BAG6</td>\n",
       "      <td>ENSG00000204463.14</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:31651656</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SLC44A4</td>\n",
       "      <td>ENSG00000204385.13</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:31865784</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.141260e+02</td>\n",
       "      <td>86</td>\n",
       "      <td>08cbd1ba-4a72-43b7-9e20-e2209ce25881</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Fusion Name Gene 1 Symbol           Gene 1 ID                                      Gene 1 Strand Gene 1 Reference Strand Gene 1 Breakpoint                              Gene 1 Input Position Gene 2 Symbol           Gene 2 ID                                      Gene 2 Strand Gene 2 Reference Strand Gene 2 Breakpoint                              Gene 2 Input Position  Breakpoint Distance  Supporting Reads                                           Read IDs    GF Program Strand Comparison       Sample Name\n",
       "23130  MT-ND4L:MT-ND4       MT-ND4L   ENSG00000212907.2                                                  +                       +        chrM:10689                                                NaN        MT-ND4   ENSG00000198886.2                                                  +                       +        chrM:10727                                                NaN         3.800000e+01               610  ff94284e-2348-4212-b7f0-50432d204860,0116346f-...  FusionSeeker        Consistent  sampleK562_pool2\n",
       "23124    PSPHP1:KMT2C        PSPHP1   ENSG00000226278.1                                                  +                       +     chr7:55773179                                                NaN         KMT2C  ENSG00000055609.21                                                  -                       -    chr7:152367259                                                NaN         9.659408e+07               147  fe0bb8e8-c947-46d5-a13e-79a50b3a59d2,0096cbc0-...  FusionSeeker        Consistent  sampleK562_pool2\n",
       "23111    BAG6:SLC44A4          BAG6  ENSG00000204463.14                                                  -                       -     chr6:31651655                                                NaN       SLC44A4  ENSG00000204385.13                                                  -                       -     chr6:31865786                                                NaN         2.141310e+05               114  ff508124-0c29-4076-bc09-5f000bd29808,000d6e61-...  FusionSeeker        Consistent  sampleK562_pool2\n",
       "12765    BAG6:SLC44A4          BAG6  ENSG00000204463.14  +,-,-,-,-,-,+,-,-,+,+,-,-,+,+,-,-,-,+,+,+,+,-,...                       -     chr6:31651657  1167-1500,38-383,37-376,28-291,35-366,30-360,1...       SLC44A4  ENSG00000204385.13  +,-,-,-,-,-,+,-,-,+,+,-,-,+,+,-,-,-,+,+,+,+,-,...                       -     chr6:31865784  80-1156,381-2334,374-1444,292-1360,367-1426,36...         2.141270e+05                93  000d6e61-632c-461e-b373-82d4fa281a8f,08cbd1ba-...        LongGF        Consistent  sampleK562_pool2\n",
       "21314    BAG6:SLC44A4          BAG6  ENSG00000204463.14                                                  -                       -     chr6:31651656                                                NaN       SLC44A4  ENSG00000204385.13                                                  -                       -     chr6:31865784                                                NaN         2.141260e+02                86               08cbd1ba-4a72-43b7-9e20-e2209ce25881        JAFFAL        Consistent  sampleK562_pool2"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize an empty list to hold the dataframes for each sample\n",
    "combined_df_list = []\n",
    "\n",
    "# Loop through each sample name and process the corresponding files\n",
    "for sample_number in sample_names:\n",
    "    LJF_filter_input_file = f'{input_sum_GF_sample_path}/summary_all_fusions.tsv'\n",
    "    LJF_df = pd.read_csv(LJF_filter_input_file, sep='\\t')\n",
    "    LJF_df['Sample Name'] = sample_number\n",
    "    combined_df_list.append(LJF_df)\n",
    "all_fusions_input_df = pd.concat(combined_df_list, ignore_index=True)\n",
    "\n",
    "\n",
    "# Drop duplicates while keeping the row with the highest 'Supporting Reads'\n",
    "all_fusions_input_df = (\n",
    "    all_fusions_input_df\n",
    "    .sort_values('Supporting Reads', ascending=False)  # Sort by 'Supporting Reads' in descending order\n",
    "    .drop_duplicates(subset=['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol', \n",
    "                              'Gene 1 Breakpoint', 'Gene 2 Breakpoint', \n",
    "                              'GF Program', 'Sample Name'], \n",
    "                     keep='first'))  # Retain the first occurrence (highest 'Supporting Reads')\n",
    "\n",
    "print(\"Total Fusions Detected: \", all_fusions_input_df.shape[0])\n",
    "\n",
    "# number of unique fusions \n",
    "unique_fusions_initial_input = all_fusions_input_df[['Fusion Name', 'Sample Name']].drop_duplicates()\n",
    "num_unique_fusions_initial_input = len(unique_fusions_initial_input)\n",
    "print('Unique Gene Fusions Initiall Detected (based on gene name): ', num_unique_fusions_initial_input)\n",
    "\n",
    "all_fusions_input_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fb700d03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Initial Detected Unique GFs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>4031</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Initial Detected Unique GFs\n",
       "0  sampleK562_pool2                         4031"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate a df to keep track of fusion counts while filtering\n",
    "fusion_counts_while_filtering = unique_fusions_initial_input.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "fusion_counts_while_filtering.columns = ['Sample Name', 'Initial Detected Unique GFs']\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8caf5e7",
   "metadata": {},
   "source": [
    "## **a.** Keep GFs with CHOP Cancer Panel Genes and Cancer Census Union of Genes  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a06436cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of CHOP genes: 119\n",
      "Number of Cancer Census Genes:  753\n",
      "Total number of combined genes: 782\n"
     ]
    }
   ],
   "source": [
    "chop_genes_df = pd.read_csv(chop_panel_genes_path, header=None)\n",
    "chop_genes = set(chop_genes_df[0].dropna())  \n",
    "print(f\"Total number of CHOP genes: {len(chop_genes)}\")\n",
    "\n",
    "###\n",
    "\n",
    "cancer_census_df = pd.read_csv(cancer_census_genes_path, sep='\\t', usecols=['GENE_SYMBOL'])\n",
    "cc_genes = set(cancer_census_df['GENE_SYMBOL'].dropna())\n",
    "print('Number of Cancer Census Genes: ', len(cc_genes))\n",
    "\n",
    "###\n",
    "\n",
    "# Combine Genes\n",
    "combined_genes_set = chop_genes.union(cc_genes)\n",
    "combined_genes_list = set(combined_genes_set)\n",
    "\n",
    "# Look at overlap\n",
    "overlap_with_chop = chop_genes.intersection(cc_genes)\n",
    "print(f\"Total number of combined genes: {len(combined_genes_list)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "10df4bfc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of fusions with at least 1 CHOP or Cancer Census gene: 581\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 1 ID</th>\n",
       "      <th>Gene 1 Strand</th>\n",
       "      <th>Gene 1 Reference Strand</th>\n",
       "      <th>Gene 1 Breakpoint</th>\n",
       "      <th>Gene 1 Input Position</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Gene 2 ID</th>\n",
       "      <th>Gene 2 Strand</th>\n",
       "      <th>Gene 2 Reference Strand</th>\n",
       "      <th>Gene 2 Breakpoint</th>\n",
       "      <th>Gene 2 Input Position</th>\n",
       "      <th>Breakpoint Distance</th>\n",
       "      <th>Supporting Reads</th>\n",
       "      <th>Read IDs</th>\n",
       "      <th>GF Program</th>\n",
       "      <th>Strand Comparison</th>\n",
       "      <th>Sample Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23124</th>\n",
       "      <td>PSPHP1:KMT2C</td>\n",
       "      <td>PSPHP1</td>\n",
       "      <td>ENSG00000226278.1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7:55773179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>ENSG00000055609.21</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:152367259</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96594080.0</td>\n",
       "      <td>147</td>\n",
       "      <td>fe0bb8e8-c947-46d5-a13e-79a50b3a59d2,0096cbc0-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10678</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290410</td>\n",
       "      <td>588-2281,39-1762,20-1940,137-389,30-1467,28-17...</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>25-587,1763-3510,1941-3680,22-136,1472-2045,17...</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23142</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21315</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>40</td>\n",
       "      <td>13581df7-a34b-4968-a43d-f984d302e89c</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23114</th>\n",
       "      <td>CFAP20DC-DT:FHIT</td>\n",
       "      <td>CFAP20DC-DT</td>\n",
       "      <td>ENSG00000243295.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:59752322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FHIT</td>\n",
       "      <td>ENSG00000189283.11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:60277452</td>\n",
       "      <td>NaN</td>\n",
       "      <td>525130.0</td>\n",
       "      <td>13</td>\n",
       "      <td>e316a59e-12e2-4b5f-97c0-eed3f519566a,0090a31e-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Fusion Name Gene 1 Symbol           Gene 1 ID                                      Gene 1 Strand Gene 1 Reference Strand Gene 1 Breakpoint                              Gene 1 Input Position Gene 2 Symbol           Gene 2 ID                                      Gene 2 Strand Gene 2 Reference Strand Gene 2 Breakpoint                              Gene 2 Input Position  Breakpoint Distance  Supporting Reads                                           Read IDs    GF Program Strand Comparison       Sample Name\n",
       "23124      PSPHP1:KMT2C        PSPHP1   ENSG00000226278.1                                                  +                       +     chr7:55773179                                                NaN         KMT2C  ENSG00000055609.21                                                  -                       -    chr7:152367259                                                NaN           96594080.0               147  fe0bb8e8-c947-46d5-a13e-79a50b3a59d2,0096cbc0-...  FusionSeeker        Consistent  sampleK562_pool2\n",
       "10678          BCR:ABL1           BCR  ENSG00000186716.21  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr22:23290410  588-2281,39-1762,20-1940,137-389,30-1467,28-17...          ABL1  ENSG00000097007.20  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr9:130854063  25-587,1763-3510,1941-3680,22-136,1472-2045,17...                  inf                52  0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...        LongGF        Consistent  sampleK562_pool2\n",
       "23142          BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290412                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854063                                                NaN                  inf                52  f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...  FusionSeeker        Consistent  sampleK562_pool2\n",
       "21315          BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854064                                                NaN                  inf                40               13581df7-a34b-4968-a43d-f984d302e89c        JAFFAL        Consistent  sampleK562_pool2\n",
       "23114  CFAP20DC-DT:FHIT   CFAP20DC-DT   ENSG00000243295.2                                                  +                       +     chr3:59752322                                                NaN          FHIT  ENSG00000189283.11                                                  -                       -     chr3:60277452                                                NaN             525130.0                13  e316a59e-12e2-4b5f-97c0-eed3f519566a,0090a31e-...  FusionSeeker        Consistent  sampleK562_pool2"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter fusions to those that involve at least one gene in the union of CHOP & Cancer Census genes\n",
    "chop_cancer_census_gfs = all_fusions_input_df[\n",
    "    all_fusions_input_df['Gene 1 Symbol'].isin(combined_genes_list) | \n",
    "    all_fusions_input_df['Gene 2 Symbol'].isin(combined_genes_list)]\n",
    "\n",
    "# Note which gene is involved\n",
    "def check_overlap(genes_set, column_name):\n",
    "    def check(row):\n",
    "        gene1_in_panel = row['Gene 1 Symbol'] in genes_set\n",
    "        gene2_in_panel = row['Gene 2 Symbol'] in genes_set\n",
    "        if gene1_in_panel and gene2_in_panel:\n",
    "            return 'Both'\n",
    "        elif gene1_in_panel:\n",
    "            return 'Gene 1'\n",
    "        elif gene2_in_panel:\n",
    "            return 'Gene 2'\n",
    "        else:\n",
    "            return 'None'\n",
    "    chop_cancer_census_gfs_copy = chop_cancer_census_gfs.copy()\n",
    "    chop_cancer_census_gfs_copy[column_name] = chop_cancer_census_gfs_copy.apply(check, axis=1)\n",
    "\n",
    "check_overlap(chop_genes, 'CHOP Fusion Panel Overlap')\n",
    "check_overlap(cc_genes, 'Cancer Census Overlap')\n",
    "\n",
    "print(f\"Number of fusions with at least 1 CHOP or Cancer Census gene: {chop_cancer_census_gfs.shape[0]}\")\n",
    "chop_cancer_census_gfs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8d241891",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Initial Detected Unique GFs</th>\n",
       "      <th>Filter on CHOP and Cancer Census Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>4031</td>\n",
       "      <td>468</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Initial Detected Unique GFs  Filter on CHOP and Cancer Census Genes\n",
       "0  sampleK562_pool2                         4031                                     468"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_chop_filter1 = chop_cancer_census_gfs.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_chop_filter1, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Filter on CHOP and Cancer Census Genes'}, inplace=True)\n",
    "fusion_counts_while_filtering['Filter on CHOP and Cancer Census Genes'] = fusion_counts_while_filtering['Filter on CHOP and Cancer Census Genes'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84b53021",
   "metadata": {},
   "source": [
    "## **b.** Keep Fusions with More than 4 Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "eff70534",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of fusions with 4 or more supporting reads: 45\n",
      "Number of fusions removed: 536\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 1 ID</th>\n",
       "      <th>Gene 1 Strand</th>\n",
       "      <th>Gene 1 Reference Strand</th>\n",
       "      <th>Gene 1 Breakpoint</th>\n",
       "      <th>Gene 1 Input Position</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Gene 2 ID</th>\n",
       "      <th>Gene 2 Strand</th>\n",
       "      <th>Gene 2 Reference Strand</th>\n",
       "      <th>Gene 2 Breakpoint</th>\n",
       "      <th>Gene 2 Input Position</th>\n",
       "      <th>Breakpoint Distance</th>\n",
       "      <th>Supporting Reads</th>\n",
       "      <th>Read IDs</th>\n",
       "      <th>GF Program</th>\n",
       "      <th>Strand Comparison</th>\n",
       "      <th>Sample Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23124</th>\n",
       "      <td>PSPHP1:KMT2C</td>\n",
       "      <td>PSPHP1</td>\n",
       "      <td>ENSG00000226278.1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7:55773179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>ENSG00000055609.21</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:152367259</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96594080.0</td>\n",
       "      <td>147</td>\n",
       "      <td>fe0bb8e8-c947-46d5-a13e-79a50b3a59d2,0096cbc0-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10678</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290410</td>\n",
       "      <td>588-2281,39-1762,20-1940,137-389,30-1467,28-17...</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>25-587,1763-3510,1941-3680,22-136,1472-2045,17...</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23142</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21315</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>40</td>\n",
       "      <td>13581df7-a34b-4968-a43d-f984d302e89c</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23114</th>\n",
       "      <td>CFAP20DC-DT:FHIT</td>\n",
       "      <td>CFAP20DC-DT</td>\n",
       "      <td>ENSG00000243295.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:59752322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FHIT</td>\n",
       "      <td>ENSG00000189283.11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:60277452</td>\n",
       "      <td>NaN</td>\n",
       "      <td>525130.0</td>\n",
       "      <td>13</td>\n",
       "      <td>e316a59e-12e2-4b5f-97c0-eed3f519566a,0090a31e-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Fusion Name Gene 1 Symbol           Gene 1 ID                                      Gene 1 Strand Gene 1 Reference Strand Gene 1 Breakpoint                              Gene 1 Input Position Gene 2 Symbol           Gene 2 ID                                      Gene 2 Strand Gene 2 Reference Strand Gene 2 Breakpoint                              Gene 2 Input Position  Breakpoint Distance  Supporting Reads                                           Read IDs    GF Program Strand Comparison       Sample Name\n",
       "23124      PSPHP1:KMT2C        PSPHP1   ENSG00000226278.1                                                  +                       +     chr7:55773179                                                NaN         KMT2C  ENSG00000055609.21                                                  -                       -    chr7:152367259                                                NaN           96594080.0               147  fe0bb8e8-c947-46d5-a13e-79a50b3a59d2,0096cbc0-...  FusionSeeker        Consistent  sampleK562_pool2\n",
       "10678          BCR:ABL1           BCR  ENSG00000186716.21  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr22:23290410  588-2281,39-1762,20-1940,137-389,30-1467,28-17...          ABL1  ENSG00000097007.20  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr9:130854063  25-587,1763-3510,1941-3680,22-136,1472-2045,17...                  inf                52  0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...        LongGF        Consistent  sampleK562_pool2\n",
       "23142          BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290412                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854063                                                NaN                  inf                52  f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...  FusionSeeker        Consistent  sampleK562_pool2\n",
       "21315          BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854064                                                NaN                  inf                40               13581df7-a34b-4968-a43d-f984d302e89c        JAFFAL        Consistent  sampleK562_pool2\n",
       "23114  CFAP20DC-DT:FHIT   CFAP20DC-DT   ENSG00000243295.2                                                  +                       +     chr3:59752322                                                NaN          FHIT  ENSG00000189283.11                                                  -                       -     chr3:60277452                                                NaN             525130.0                13  e316a59e-12e2-4b5f-97c0-eed3f519566a,0090a31e-...  FusionSeeker        Consistent  sampleK562_pool2"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove those with less than 4 supporting reads\n",
    "read_filter_df = chop_cancer_census_gfs[chop_cancer_census_gfs['Supporting Reads'] >= set_read_support]\n",
    "\n",
    "# Number of entries removed\n",
    "num_removed_read_filter = chop_cancer_census_gfs.shape[0] - read_filter_df.shape[0]\n",
    "\n",
    "print(\"Number of fusions with 4 or more supporting reads:\", read_filter_df.shape[0])\n",
    "print(\"Number of fusions removed:\", num_removed_read_filter)\n",
    "read_filter_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c3f98771",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Initial Detected Unique GFs</th>\n",
       "      <th>Filter on CHOP and Cancer Census Genes</th>\n",
       "      <th>Filter on Read Support</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>4031</td>\n",
       "      <td>468</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Initial Detected Unique GFs  Filter on CHOP and Cancer Census Genes  Filter on Read Support\n",
       "0  sampleK562_pool2                         4031                                     468                      26"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_read_filter2 = read_filter_df.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_read_filter2, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Filter on Read Support'}, inplace=True)\n",
    "fusion_counts_while_filtering['Filter on Read Support'] = fusion_counts_while_filtering['Filter on Read Support'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3d3136d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d0b4e5d4",
   "metadata": {},
   "source": [
    "## **c.** Remove Fusions with Unrelated Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2d14721a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of fusions remaining: 36\n",
      "Number of fusions removed: 9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 1 ID</th>\n",
       "      <th>Gene 1 Strand</th>\n",
       "      <th>Gene 1 Reference Strand</th>\n",
       "      <th>Gene 1 Breakpoint</th>\n",
       "      <th>Gene 1 Input Position</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Gene 2 ID</th>\n",
       "      <th>Gene 2 Strand</th>\n",
       "      <th>Gene 2 Reference Strand</th>\n",
       "      <th>Gene 2 Breakpoint</th>\n",
       "      <th>Gene 2 Input Position</th>\n",
       "      <th>Breakpoint Distance</th>\n",
       "      <th>Supporting Reads</th>\n",
       "      <th>Read IDs</th>\n",
       "      <th>GF Program</th>\n",
       "      <th>Strand Comparison</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Gene 1 Gene Type</th>\n",
       "      <th>Gene 2 Gene Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290410</td>\n",
       "      <td>588-2281,39-1762,20-1940,137-389,30-1467,28-17...</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>25-587,1763-3510,1941-3680,22-136,1472-2045,17...</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>40</td>\n",
       "      <td>13581df7-a34b-4968-a43d-f984d302e89c</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CFAP20DC-DT:FHIT</td>\n",
       "      <td>CFAP20DC-DT</td>\n",
       "      <td>ENSG00000243295.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:59752322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FHIT</td>\n",
       "      <td>ENSG00000189283.11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:60277452</td>\n",
       "      <td>NaN</td>\n",
       "      <td>525130.0</td>\n",
       "      <td>13</td>\n",
       "      <td>e316a59e-12e2-4b5f-97c0-eed3f519566a,0090a31e-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NUP214:XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+,+,-,-,-,-,+,+,-,-</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199013</td>\n",
       "      <td>31-118,32-1068,1044-2974,1141-1269,1165-2437,2...</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-,-,+,+,+,+,-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>116-2661,1066-3601,57-1046,42-1143,62-1167,48-...</td>\n",
       "      <td>inf</td>\n",
       "      <td>10</td>\n",
       "      <td>002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XKR3:NUP214</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>9</td>\n",
       "      <td>e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CFAP20DC-DT:FHIT</td>\n",
       "      <td>CFAP20DC-DT</td>\n",
       "      <td>ENSG00000243295.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:59752322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FHIT</td>\n",
       "      <td>ENSG00000189283.11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:60215079</td>\n",
       "      <td>NaN</td>\n",
       "      <td>462757.0</td>\n",
       "      <td>9</td>\n",
       "      <td>f2ab4ec1-1556-47a9-bacf-01b85d423941,0951fff1-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NUP214:XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808083</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>9</td>\n",
       "      <td>24016df2-c090-4de4-8fff-80c7d29648c9</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130780368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>7</td>\n",
       "      <td>c17e7c8e-0149-4d8b-9237-26b5858ff694,03d7adcb-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CFAP20DC-DT:FHIT</td>\n",
       "      <td>CFAP20DC-DT</td>\n",
       "      <td>ENSG00000243295.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:59752322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FHIT</td>\n",
       "      <td>ENSG00000189283.11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:60178440</td>\n",
       "      <td>NaN</td>\n",
       "      <td>426118.0</td>\n",
       "      <td>6</td>\n",
       "      <td>fa95c101-4709-49a2-8673-dbfc2053bd2d,4dcf3b48-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CFAP20DC-DT:FHIT</td>\n",
       "      <td>CFAP20DC-DT</td>\n",
       "      <td>ENSG00000243295.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:59752324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FHIT</td>\n",
       "      <td>ENSG00000189283.11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:61200616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1448292.0</td>\n",
       "      <td>5</td>\n",
       "      <td>e759d5ca-f7df-48fa-9efc-5c2b594e9cb6,48245748-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TMEM9:MAPK1</td>\n",
       "      <td>TMEM9</td>\n",
       "      <td>ENSG00000116857.17</td>\n",
       "      <td>-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:201135722</td>\n",
       "      <td>45-571,40-560,401-920,366-889</td>\n",
       "      <td>MAPK1</td>\n",
       "      <td>ENSG00000100030.15</td>\n",
       "      <td>-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:21760659</td>\n",
       "      <td>569-910,558-893,67-403,38-368</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>4acc5224-f5e3-4d7e-9c6e-5388166c0913,554b3f0f-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ATP5PD:GATA2</td>\n",
       "      <td>ATP5PD</td>\n",
       "      <td>ENSG00000167863.12</td>\n",
       "      <td>+,-,+,-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17:75038955</td>\n",
       "      <td>536-984,63-482,538-979,32-479</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>ENSG00000179348.13</td>\n",
       "      <td>+,-,+,-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:128479921</td>\n",
       "      <td>56-541,477-962,44-543,474-958</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>171e4ad8-b4f3-4a1a-9f6d-b94951e22650,3b7a1e59-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ATIC:MYL6</td>\n",
       "      <td>ATIC</td>\n",
       "      <td>ENSG00000138363.15</td>\n",
       "      <td>+,-,-,+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:215338900</td>\n",
       "      <td>312-1596,396-1655,421-1706,31-1312</td>\n",
       "      <td>MYL6</td>\n",
       "      <td>ENSG00000092841.19</td>\n",
       "      <td>+,-,-,+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr12:56160054</td>\n",
       "      <td>1595-1961,42-395,60-422,1311-1665</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>44d5b97c-7af2-40f5-aa64-099169213edd,712c8cbf-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PAFAH1B2:GGPS1</td>\n",
       "      <td>PAFAH1B2</td>\n",
       "      <td>ENSG00000168092.14</td>\n",
       "      <td>-,-,-,+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr11:117161256</td>\n",
       "      <td>1041-1433,1023-1425,1044-1457,29-432</td>\n",
       "      <td>GGPS1</td>\n",
       "      <td>ENSG00000152904.11</td>\n",
       "      <td>-,-,-,+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:235343520</td>\n",
       "      <td>66-1041,41-1027,50-1048,428-1440</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>4a2a0e57-d003-4bcc-805f-f7cb0d80c8e1,713857e5-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NACA:TBL1XR1</td>\n",
       "      <td>NACA</td>\n",
       "      <td>ENSG00000196531.14</td>\n",
       "      <td>+,+,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12:56712813</td>\n",
       "      <td>268-913,281-914,270-837,289-895</td>\n",
       "      <td>TBL1XR1</td>\n",
       "      <td>ENSG00000177565.19</td>\n",
       "      <td>+,+,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:177020989</td>\n",
       "      <td>35-268,54-281,42-270,55-289</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>0113f114-ea1d-46b8-9f12-b8927fb70526,145733cb-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NF2:C9orf78</td>\n",
       "      <td>NF2</td>\n",
       "      <td>ENSG00000186575.20</td>\n",
       "      <td>-,-,+,+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:29696188</td>\n",
       "      <td>580-716,582-726,31-175,30-171</td>\n",
       "      <td>C9orf78</td>\n",
       "      <td>ENSG00000136819.15</td>\n",
       "      <td>+,+,-,-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9:129827841</td>\n",
       "      <td>45-588,65-590,167-719,163-697</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>7fed01b8-82c3-42de-b2a0-4aedc8b79e62,8513da81-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NACA:TBL1XR1</td>\n",
       "      <td>NACA</td>\n",
       "      <td>ENSG00000196531.14</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12:56712813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TBL1XR1</td>\n",
       "      <td>ENSG00000177565.19</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:177020989</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>e1e4fef6-8bcb-4f5c-825c-603099a328a5,0113f114-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MALAT1:AP2S1</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>ENSG00000251562.11</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr11:65500724</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>ENSG00000042753.12</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr19:46838367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>baad0435-ae78-4a67-a97c-ac7ce8007982,1b711182-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>KCTD2:GATA2</td>\n",
       "      <td>KCTD2</td>\n",
       "      <td>ENSG00000180901.11</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17:75038955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>ENSG00000179348.13</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:128479921</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>8f53400a-43c2-406a-b814-0c6258463263,171e4ad8-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TFRC:SBNO1</td>\n",
       "      <td>TFRC</td>\n",
       "      <td>ENSG00000072274.14</td>\n",
       "      <td>-,-,+,-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:196075210</td>\n",
       "      <td>28-276,71-286,349-607,71-284</td>\n",
       "      <td>SBNO1</td>\n",
       "      <td>ENSG00000139697.14</td>\n",
       "      <td>-,-,+,-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12:123289422</td>\n",
       "      <td>270-571,280-593,47-355,278-595</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>230b115b-a161-455a-9476-3c296939eb72,6fa7cb8c-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MAX:PABPC1</td>\n",
       "      <td>MAX</td>\n",
       "      <td>ENSG00000125952.21</td>\n",
       "      <td>-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr14:65084191</td>\n",
       "      <td>31-500,32-491,210-664,225-676</td>\n",
       "      <td>PABPC1</td>\n",
       "      <td>ENSG00000070756.17</td>\n",
       "      <td>-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr8:100703094</td>\n",
       "      <td>492-674,482-664,44-215,56-234</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>49b2c69e-6122-40ec-9927-61ef91be006c,a6d52ae1-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SBNO1:TFRC</td>\n",
       "      <td>SBNO1</td>\n",
       "      <td>ENSG00000139697.14</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12:123289422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TFRC</td>\n",
       "      <td>ENSG00000072274.14</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:196075210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>dffe7771-6ea7-4384-b778-fa41014ff557,230b115b-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CFAP20DC-DT:FHIT</td>\n",
       "      <td>CFAP20DC-DT</td>\n",
       "      <td>ENSG00000243295.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:59752324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FHIT</td>\n",
       "      <td>ENSG00000189283.11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:61042043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1289719.0</td>\n",
       "      <td>4</td>\n",
       "      <td>a17df4eb-30b5-41c6-818d-858ad27e25d7,43ed4669-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NF2:C9orf78</td>\n",
       "      <td>NF2</td>\n",
       "      <td>ENSG00000186575.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:29696188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C9orf78</td>\n",
       "      <td>ENSG00000136819.15</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9:129827841</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>ffd822f2-7f78-4d44-8df0-d0eb6bf695e3,7fed01b8-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MAX:PABPC1</td>\n",
       "      <td>MAX</td>\n",
       "      <td>ENSG00000125952.21</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr14:65084190</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PABPC1</td>\n",
       "      <td>ENSG00000070756.17</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr8:100703095</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>f019ab0b-7125-435e-a473-19eacc0d3b75,49b2c69e-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>GGPS1:PAFAH1B2</td>\n",
       "      <td>GGPS1</td>\n",
       "      <td>ENSG00000152904.11</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:235343520</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PAFAH1B2</td>\n",
       "      <td>ENSG00000168092.14</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr11:117161257</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>f4c72fd1-f9da-43a3-9690-1afe7c139ed7,4a2a0e57-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>TMEM9:MAPK1</td>\n",
       "      <td>TMEM9</td>\n",
       "      <td>ENSG00000116857.17</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:201135722</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAPK1</td>\n",
       "      <td>ENSG00000100030.15</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:21760659</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>7cba2e33-60eb-4583-8c44-630b2571e04e,4acc5224-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>PAFAH1B2:GGPS1</td>\n",
       "      <td>PAFAH1B2</td>\n",
       "      <td>ENSG00000168092.14</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr11:117161253</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGPS1</td>\n",
       "      <td>ENSG00000152904.11</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:235343527</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>713857e5-85b9-4767-b6b8-0da5023f502f</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MALAT1:AP2S1</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>ENSG00000251562.11</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr11:65500719</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>ENSG00000042753.12</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr19:46838361</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>1b711182-fe6e-42e7-b208-c5eff0b94cfd</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>MAP3K13:ATP5O</td>\n",
       "      <td>MAP3K13</td>\n",
       "      <td>ENSG00000073803.14</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:185417606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATP5O</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr21:33903970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>91f1eebe-5222-4478-9fe4-ad1e69fbf020</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Inconsistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>TMEM9:MAPK1</td>\n",
       "      <td>TMEM9</td>\n",
       "      <td>ENSG00000116857.17</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:201135726</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAPK1</td>\n",
       "      <td>ENSG00000100030.15</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:21760656</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>6e262750-c2f6-4bbb-9e10-6403035f9605</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>ATIC:MYL6</td>\n",
       "      <td>ATIC</td>\n",
       "      <td>ENSG00000138363.15</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:215338899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MYL6</td>\n",
       "      <td>ENSG00000092841.19</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr12:56160057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>44d5b97c-7af2-40f5-aa64-099169213edd</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NF2:C9orf78</td>\n",
       "      <td>NF2</td>\n",
       "      <td>ENSG00000186575.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:29696179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C9orf78</td>\n",
       "      <td>ENSG00000136819.15</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9:129827833</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>e746949c-b948-40fe-9252-4346aae67dfc</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>TFRC:SBNO1</td>\n",
       "      <td>TFRC</td>\n",
       "      <td>ENSG00000072274.14</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:196075219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SBNO1</td>\n",
       "      <td>ENSG00000139697.14</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12:123289416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>230b115b-a161-455a-9476-3c296939eb72</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>MYL6:ATIC</td>\n",
       "      <td>MYL6</td>\n",
       "      <td>ENSG00000092841.19</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr12:56160054</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATIC</td>\n",
       "      <td>ENSG00000138363.15</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:215338901</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>4</td>\n",
       "      <td>7dbe40ab-8406-4a70-b345-abd27d49b811,44d5b97c-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Fusion Name Gene 1 Symbol           Gene 1 ID                                      Gene 1 Strand Gene 1 Reference Strand Gene 1 Breakpoint                              Gene 1 Input Position Gene 2 Symbol           Gene 2 ID                                      Gene 2 Strand Gene 2 Reference Strand Gene 2 Breakpoint                              Gene 2 Input Position  Breakpoint Distance  Supporting Reads                                           Read IDs    GF Program Strand Comparison       Sample Name Gene 1 Gene Type Gene 2 Gene Type\n",
       "1           BCR:ABL1           BCR  ENSG00000186716.21  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr22:23290410  588-2281,39-1762,20-1940,137-389,30-1467,28-17...          ABL1  ENSG00000097007.20  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr9:130854063  25-587,1763-3510,1941-3680,22-136,1472-2045,17...                  inf                52  0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "2           BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290412                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854063                                                NaN                  inf                52  f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "3           BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854064                                                NaN                  inf                40               13581df7-a34b-4968-a43d-f984d302e89c        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "4   CFAP20DC-DT:FHIT   CFAP20DC-DT   ENSG00000243295.2                                                  +                       +     chr3:59752322                                                NaN          FHIT  ENSG00000189283.11                                                  -                       -     chr3:60277452                                                NaN             525130.0                13  e316a59e-12e2-4b5f-97c0-eed3f519566a,0090a31e-...  FusionSeeker        Consistent  sampleK562_pool2           lncRNA   protein_coding\n",
       "6        NUP214:XKR3        NUP214  ENSG00000126883.19                                +,+,-,-,-,-,+,+,-,-                       +    chr9:131199013  31-118,32-1068,1044-2974,1141-1269,1165-2437,2...          XKR3   ENSG00000172967.8                                -,-,+,+,+,+,-,-,+,+                       -    chr22:16808085  116-2661,1066-3601,57-1046,42-1143,62-1167,48-...                  inf                10  002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "7        XKR3:NUP214          XKR3   ENSG00000172967.8                                                  -                       -    chr22:16808085                                                NaN        NUP214  ENSG00000126883.19                                                  +                       +    chr9:131199011                                                NaN                  inf                 9  e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "8   CFAP20DC-DT:FHIT   CFAP20DC-DT   ENSG00000243295.2                                                  +                       +     chr3:59752322                                                NaN          FHIT  ENSG00000189283.11                                                  -                       -     chr3:60215079                                                NaN             462757.0                 9  f2ab4ec1-1556-47a9-bacf-01b85d423941,0951fff1-...  FusionSeeker        Consistent  sampleK562_pool2           lncRNA   protein_coding\n",
       "9        NUP214:XKR3        NUP214  ENSG00000126883.19                                                  +                       +    chr9:131199015                                                NaN          XKR3   ENSG00000172967.8                                                  -                       -    chr22:16808083                                                NaN                  inf                 9               24016df2-c090-4de4-8fff-80c7d29648c9        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "12          BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130780368                                                NaN                  inf                 7  c17e7c8e-0149-4d8b-9237-26b5858ff694,03d7adcb-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "13  CFAP20DC-DT:FHIT   CFAP20DC-DT   ENSG00000243295.2                                                  +                       +     chr3:59752322                                                NaN          FHIT  ENSG00000189283.11                                                  -                       -     chr3:60178440                                                NaN             426118.0                 6  fa95c101-4709-49a2-8673-dbfc2053bd2d,4dcf3b48-...  FusionSeeker        Consistent  sampleK562_pool2           lncRNA   protein_coding\n",
       "15  CFAP20DC-DT:FHIT   CFAP20DC-DT   ENSG00000243295.2                                                  +                       +     chr3:59752324                                                NaN          FHIT  ENSG00000189283.11                                                  -                       -     chr3:61200616                                                NaN            1448292.0                 5  e759d5ca-f7df-48fa-9efc-5c2b594e9cb6,48245748-...  FusionSeeker        Consistent  sampleK562_pool2           lncRNA   protein_coding\n",
       "18       TMEM9:MAPK1         TMEM9  ENSG00000116857.17                                            -,-,+,+                       -    chr1:201135722                      45-571,40-560,401-920,366-889         MAPK1  ENSG00000100030.15                                            -,-,+,+                       -    chr22:21760659                      569-910,558-893,67-403,38-368                  inf                 4  4acc5224-f5e3-4d7e-9c6e-5388166c0913,554b3f0f-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "19      ATP5PD:GATA2        ATP5PD  ENSG00000167863.12                                            +,-,+,-                       -    chr17:75038955                      536-984,63-482,538-979,32-479         GATA2  ENSG00000179348.13                                            +,-,+,-                       -    chr3:128479921                      56-541,477-962,44-543,474-958                  inf                 4  171e4ad8-b4f3-4a1a-9f6d-b94951e22650,3b7a1e59-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "20         ATIC:MYL6          ATIC  ENSG00000138363.15                                            +,-,-,+                       +    chr2:215338900                 312-1596,396-1655,421-1706,31-1312          MYL6  ENSG00000092841.19                                            +,-,-,+                       +    chr12:56160054                  1595-1961,42-395,60-422,1311-1665                  inf                 4  44d5b97c-7af2-40f5-aa64-099169213edd,712c8cbf-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "21    PAFAH1B2:GGPS1      PAFAH1B2  ENSG00000168092.14                                            -,-,-,+                       +   chr11:117161256               1041-1433,1023-1425,1044-1457,29-432         GGPS1  ENSG00000152904.11                                            -,-,-,+                       +    chr1:235343520                   66-1041,41-1027,50-1048,428-1440                  inf                 4  4a2a0e57-d003-4bcc-805f-f7cb0d80c8e1,713857e5-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "22      NACA:TBL1XR1          NACA  ENSG00000196531.14                                            +,+,+,+                       -    chr12:56712813                    268-913,281-914,270-837,289-895       TBL1XR1  ENSG00000177565.19                                            +,+,+,+                       -    chr3:177020989                        35-268,54-281,42-270,55-289                  inf                 4  0113f114-ea1d-46b8-9f12-b8927fb70526,145733cb-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "23       NF2:C9orf78           NF2  ENSG00000186575.20                                            -,-,+,+                       +    chr22:29696188                      580-716,582-726,31-175,30-171       C9orf78  ENSG00000136819.15                                            +,+,-,-                       -    chr9:129827841                      45-588,65-590,167-719,163-697                  inf                 4  7fed01b8-82c3-42de-b2a0-4aedc8b79e62,8513da81-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "24      NACA:TBL1XR1          NACA  ENSG00000196531.14                                                  -                       -    chr12:56712813                                                NaN       TBL1XR1  ENSG00000177565.19                                                  -                       -    chr3:177020989                                                NaN                  inf                 4  e1e4fef6-8bcb-4f5c-825c-603099a328a5,0113f114-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "26      MALAT1:AP2S1        MALAT1  ENSG00000251562.11                                                  +                       +    chr11:65500724                                                NaN         AP2S1  ENSG00000042753.12                                                  -                       -    chr19:46838367                                                NaN                  inf                 4  baad0435-ae78-4a67-a97c-ac7ce8007982,1b711182-...  FusionSeeker        Consistent  sampleK562_pool2           lncRNA   protein_coding\n",
       "27       KCTD2:GATA2         KCTD2  ENSG00000180901.11                                                  +                       +    chr17:75038955                                                NaN         GATA2  ENSG00000179348.13                                                  -                       -    chr3:128479921                                                NaN                  inf                 4  8f53400a-43c2-406a-b814-0c6258463263,171e4ad8-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "29        TFRC:SBNO1          TFRC  ENSG00000072274.14                                            -,-,+,-                       -    chr3:196075210                       28-276,71-286,349-607,71-284         SBNO1  ENSG00000139697.14                                            -,-,+,-                       -   chr12:123289422                     270-571,280-593,47-355,278-595                  inf                 4  230b115b-a161-455a-9476-3c296939eb72,6fa7cb8c-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "30        MAX:PABPC1           MAX  ENSG00000125952.21                                            -,-,+,+                       -    chr14:65084191                      31-500,32-491,210-664,225-676        PABPC1  ENSG00000070756.17                                            -,-,+,+                       -    chr8:100703094                      492-674,482-664,44-215,56-234                  inf                 4  49b2c69e-6122-40ec-9927-61ef91be006c,a6d52ae1-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "31        SBNO1:TFRC         SBNO1  ENSG00000139697.14                                                  -                       -   chr12:123289422                                                NaN          TFRC  ENSG00000072274.14                                                  -                       -    chr3:196075210                                                NaN                  inf                 4  dffe7771-6ea7-4384-b778-fa41014ff557,230b115b-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "32  CFAP20DC-DT:FHIT   CFAP20DC-DT   ENSG00000243295.2                                                  +                       +     chr3:59752324                                                NaN          FHIT  ENSG00000189283.11                                                  -                       -     chr3:61042043                                                NaN            1289719.0                 4  a17df4eb-30b5-41c6-818d-858ad27e25d7,43ed4669-...  FusionSeeker        Consistent  sampleK562_pool2           lncRNA   protein_coding\n",
       "33       NF2:C9orf78           NF2  ENSG00000186575.20                                                  +                       +    chr22:29696188                                                NaN       C9orf78  ENSG00000136819.15                                                  -                       -    chr9:129827841                                                NaN                  inf                 4  ffd822f2-7f78-4d44-8df0-d0eb6bf695e3,7fed01b8-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "34        MAX:PABPC1           MAX  ENSG00000125952.21                                                  -                       -    chr14:65084190                                                NaN        PABPC1  ENSG00000070756.17                                                  -                       -    chr8:100703095                                                NaN                  inf                 4  f019ab0b-7125-435e-a473-19eacc0d3b75,49b2c69e-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "35    GGPS1:PAFAH1B2         GGPS1  ENSG00000152904.11                                                  +                       +    chr1:235343520                                                NaN      PAFAH1B2  ENSG00000168092.14                                                  +                       +   chr11:117161257                                                NaN                  inf                 4  f4c72fd1-f9da-43a3-9690-1afe7c139ed7,4a2a0e57-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "36       TMEM9:MAPK1         TMEM9  ENSG00000116857.17                                                  -                       -    chr1:201135722                                                NaN         MAPK1  ENSG00000100030.15                                                  -                       -    chr22:21760659                                                NaN                  inf                 4  7cba2e33-60eb-4583-8c44-630b2571e04e,4acc5224-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "37    PAFAH1B2:GGPS1      PAFAH1B2  ENSG00000168092.14                                                  +                       +   chr11:117161253                                                NaN         GGPS1  ENSG00000152904.11                                                  +                       +    chr1:235343527                                                NaN                  inf                 4               713857e5-85b9-4767-b6b8-0da5023f502f        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "38      MALAT1:AP2S1        MALAT1  ENSG00000251562.11                                                  +                       +    chr11:65500719                                                NaN         AP2S1  ENSG00000042753.12                                                  -                       -    chr19:46838361                                                NaN                  inf                 4               1b711182-fe6e-42e7-b208-c5eff0b94cfd        JAFFAL        Consistent  sampleK562_pool2           lncRNA   protein_coding\n",
       "39     MAP3K13:ATP5O       MAP3K13  ENSG00000073803.14                                                  +                       +    chr3:185417606                                                NaN         ATP5O                 NaN                                                  -                     NaN    chr21:33903970                                                NaN                  inf                 4               91f1eebe-5222-4478-9fe4-ad1e69fbf020        JAFFAL      Inconsistent  sampleK562_pool2   protein_coding          Unknown\n",
       "40       TMEM9:MAPK1         TMEM9  ENSG00000116857.17                                                  -                       -    chr1:201135726                                                NaN         MAPK1  ENSG00000100030.15                                                  -                       -    chr22:21760656                                                NaN                  inf                 4               6e262750-c2f6-4bbb-9e10-6403035f9605        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "41         ATIC:MYL6          ATIC  ENSG00000138363.15                                                  +                       +    chr2:215338899                                                NaN          MYL6  ENSG00000092841.19                                                  +                       +    chr12:56160057                                                NaN                  inf                 4               44d5b97c-7af2-40f5-aa64-099169213edd        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "42       NF2:C9orf78           NF2  ENSG00000186575.20                                                  +                       +    chr22:29696179                                                NaN       C9orf78  ENSG00000136819.15                                                  -                       -    chr9:129827833                                                NaN                  inf                 4               e746949c-b948-40fe-9252-4346aae67dfc        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "43        TFRC:SBNO1          TFRC  ENSG00000072274.14                                                  -                       -    chr3:196075219                                                NaN         SBNO1  ENSG00000139697.14                                                  -                       -   chr12:123289416                                                NaN                  inf                 4               230b115b-a161-455a-9476-3c296939eb72        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "44         MYL6:ATIC          MYL6  ENSG00000092841.19                                                  +                       +    chr12:56160054                                                NaN          ATIC  ENSG00000138363.15                                                  +                       +    chr2:215338901                                                NaN                  inf                 4  7dbe40ab-8406-4a70-b345-abd27d49b811,44d5b97c-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove Pseudogenes\n",
    "\n",
    "# Filter rows where 'feature' column contains the word 'gene'\n",
    "gene_annotation_ref_df = annotation_ref_df[annotation_ref_df['feature'] == 'gene']\n",
    "\n",
    "# Merge Gene 1 gene type\n",
    "def merge_gene_type(df, gene_col, gene_annotation_ref_df, suffix):\n",
    "    merged_df = df.merge(gene_annotation_ref_df[['gene_id', 'gene_type']], \n",
    "                         left_on=gene_col, \n",
    "                         right_on='gene_id', \n",
    "                         how='left')\n",
    "    merged_df[f'{suffix} Gene Type'] = merged_df['gene_type'].fillna('Unknown')\n",
    "    merged_df.drop(columns=['gene_type', 'gene_id'], inplace=True)\n",
    "    return merged_df\n",
    "\n",
    "# Merge Gene 1 data\n",
    "gene1_merged = merge_gene_type(read_filter_df, 'Gene 1 ID', gene_annotation_ref_df, 'Gene 1')\n",
    "\n",
    "# Merge Gene 2 data\n",
    "filtered_gene_types_added = merge_gene_type(gene1_merged, 'Gene 2 ID', gene_annotation_ref_df, 'Gene 2')\n",
    "\n",
    "# Filter for protein-coding genes and exclude pseudogenes\n",
    "filtered_gene_types_protein_coding = filtered_gene_types_added[\n",
    "    (filtered_gene_types_added['Gene 1 Gene Type'] == 'protein_coding') | \n",
    "    (filtered_gene_types_added['Gene 2 Gene Type'] == 'protein_coding')]\n",
    "\n",
    "filtered_gene_types_protein_coding = filtered_gene_types_protein_coding[\n",
    "    ~filtered_gene_types_protein_coding['Gene 1 Gene Type'].str.contains('pseudogene', na=False) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 2 Gene Type'].str.contains('pseudogene', na=False)]\n",
    "\n",
    "\n",
    "# Additionally, remove GFs: # where either 'Gene 1 Breakpoint' or 'Gene 2 Breakpoint' contains 'chrM'\n",
    "# or 'Gene 1 Symbol' or 'Gene 2 Symbol' starts with 'RPL', 'MRP', 'RPS', or 'HLA'\n",
    "\n",
    "# Patterns to filter out\n",
    "chromosome_pattern = ['chrM']\n",
    "ribo_pattern = r'^(RP|RPL|MRP|RPS)'  # Pattern for ribosomal proteins\n",
    "hla_pattern = r'^(HLA)'  # Pattern for HLA genes\n",
    "\n",
    "# Perform the combined filtering to remove unwanted GFs\n",
    "filtered_no_unrelated = filtered_gene_types_protein_coding[\n",
    "    ~filtered_gene_types_protein_coding['Gene 1 Breakpoint'].str.contains('|'.join(chromosome_pattern), na=False) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 2 Breakpoint'].str.contains('|'.join(chromosome_pattern), na=False) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 1 Symbol'].str.match(ribo_pattern) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 2 Symbol'].str.match(ribo_pattern) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 1 Symbol'].str.match(hla_pattern) & \n",
    "    ~filtered_gene_types_protein_coding['Gene 2 Symbol'].str.match(hla_pattern)\n",
    "]\n",
    "\n",
    "# Display the filtered results\n",
    "print(\"Number of fusions remaining:\", filtered_no_unrelated.shape[0])\n",
    "print(\"Number of fusions removed:\", read_filter_df.shape[0] - filtered_no_unrelated.shape[0])\n",
    "filtered_no_unrelated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bef37f80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Initial Detected Unique GFs</th>\n",
       "      <th>Filter on CHOP and Cancer Census Genes</th>\n",
       "      <th>Filter on Read Support</th>\n",
       "      <th>Filter on Unrelated Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>4031</td>\n",
       "      <td>468</td>\n",
       "      <td>26</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Initial Detected Unique GFs  Filter on CHOP and Cancer Census Genes  Filter on Read Support  Filter on Unrelated Genes\n",
       "0  sampleK562_pool2                         4031                                     468                      26                         18"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_gene_type_filter3 = filtered_no_unrelated.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_gene_type_filter3, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Filter on Unrelated Genes'}, inplace=True)\n",
    "fusion_counts_while_filtering['Filter on Unrelated Genes'] = fusion_counts_while_filtering['Filter on Unrelated Genes'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b9de4b4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "895bb166",
   "metadata": {},
   "source": [
    "##  **d.** Keep Fusions with Consistent Strandness   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dccc1132",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of fusions with Consistent Strandness: 35\n",
      "Number of fusions removed: 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 1 ID</th>\n",
       "      <th>Gene 1 Strand</th>\n",
       "      <th>Gene 1 Reference Strand</th>\n",
       "      <th>Gene 1 Breakpoint</th>\n",
       "      <th>Gene 1 Input Position</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Gene 2 ID</th>\n",
       "      <th>Gene 2 Strand</th>\n",
       "      <th>Gene 2 Reference Strand</th>\n",
       "      <th>Gene 2 Breakpoint</th>\n",
       "      <th>Gene 2 Input Position</th>\n",
       "      <th>Breakpoint Distance</th>\n",
       "      <th>Supporting Reads</th>\n",
       "      <th>Read IDs</th>\n",
       "      <th>GF Program</th>\n",
       "      <th>Strand Comparison</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Gene 1 Gene Type</th>\n",
       "      <th>Gene 2 Gene Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290410</td>\n",
       "      <td>588-2281,39-1762,20-1940,137-389,30-1467,28-17...</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>25-587,1763-3510,1941-3680,22-136,1472-2045,17...</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>40</td>\n",
       "      <td>13581df7-a34b-4968-a43d-f984d302e89c</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CFAP20DC-DT:FHIT</td>\n",
       "      <td>CFAP20DC-DT</td>\n",
       "      <td>ENSG00000243295.2</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3:59752322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FHIT</td>\n",
       "      <td>ENSG00000189283.11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3:60277452</td>\n",
       "      <td>NaN</td>\n",
       "      <td>525130.0</td>\n",
       "      <td>13</td>\n",
       "      <td>e316a59e-12e2-4b5f-97c0-eed3f519566a,0090a31e-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NUP214:XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+,+,-,-,-,-,+,+,-,-</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199013</td>\n",
       "      <td>31-118,32-1068,1044-2974,1141-1269,1165-2437,2...</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-,-,+,+,+,+,-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>116-2661,1066-3601,57-1046,42-1143,62-1167,48-...</td>\n",
       "      <td>inf</td>\n",
       "      <td>10</td>\n",
       "      <td>002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Fusion Name Gene 1 Symbol           Gene 1 ID                                      Gene 1 Strand Gene 1 Reference Strand Gene 1 Breakpoint                              Gene 1 Input Position Gene 2 Symbol           Gene 2 ID                                      Gene 2 Strand Gene 2 Reference Strand Gene 2 Breakpoint                              Gene 2 Input Position  Breakpoint Distance  Supporting Reads                                           Read IDs    GF Program Strand Comparison       Sample Name Gene 1 Gene Type Gene 2 Gene Type\n",
       "1          BCR:ABL1           BCR  ENSG00000186716.21  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr22:23290410  588-2281,39-1762,20-1940,137-389,30-1467,28-17...          ABL1  ENSG00000097007.20  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr9:130854063  25-587,1763-3510,1941-3680,22-136,1472-2045,17...                  inf                52  0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "2          BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290412                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854063                                                NaN                  inf                52  f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "3          BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854064                                                NaN                  inf                40               13581df7-a34b-4968-a43d-f984d302e89c        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "4  CFAP20DC-DT:FHIT   CFAP20DC-DT   ENSG00000243295.2                                                  +                       +     chr3:59752322                                                NaN          FHIT  ENSG00000189283.11                                                  -                       -     chr3:60277452                                                NaN             525130.0                13  e316a59e-12e2-4b5f-97c0-eed3f519566a,0090a31e-...  FusionSeeker        Consistent  sampleK562_pool2           lncRNA   protein_coding\n",
       "6       NUP214:XKR3        NUP214  ENSG00000126883.19                                +,+,-,-,-,-,+,+,-,-                       +    chr9:131199013  31-118,32-1068,1044-2974,1141-1269,1165-2437,2...          XKR3   ENSG00000172967.8                                -,-,+,+,+,+,-,-,+,+                       -    chr22:16808085  116-2661,1066-3601,57-1046,42-1143,62-1167,48-...                  inf                10  002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove those with inconsistent strandness \n",
    "consistent_filter_df = filtered_no_unrelated[filtered_no_unrelated['Strand Comparison'] == 'Consistent']\n",
    "\n",
    "# Calculate how many entries were removed\n",
    "num_removed_consistent_filter = filtered_no_unrelated.shape[0] - consistent_filter_df.shape[0]\n",
    "\n",
    "# Display the results\n",
    "print('Number of fusions with Consistent Strandness:', consistent_filter_df.shape[0])\n",
    "print(\"Number of fusions removed:\", num_removed_consistent_filter)\n",
    "consistent_filter_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "45093a8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Initial Detected Unique GFs</th>\n",
       "      <th>Filter on CHOP and Cancer Census Genes</th>\n",
       "      <th>Filter on Read Support</th>\n",
       "      <th>Filter on Unrelated Genes</th>\n",
       "      <th>Filter on Consistent Strandness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>4031</td>\n",
       "      <td>468</td>\n",
       "      <td>26</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Initial Detected Unique GFs  Filter on CHOP and Cancer Census Genes  Filter on Read Support  Filter on Unrelated Genes  Filter on Consistent Strandness\n",
       "0  sampleK562_pool2                         4031                                     468                      26                         18                               17"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_consistent_filter4 = consistent_filter_df.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_consistent_filter4, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Filter on Consistent Strandness'}, inplace=True)\n",
    "fusion_counts_while_filtering['Filter on Consistent Strandness'] = fusion_counts_while_filtering['Filter on Consistent Strandness'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c461d2a4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b09a9009",
   "metadata": {},
   "source": [
    "##  **e.** Remove Fusions Detected in Healthy GTEx Brain Tissue Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "386c0c92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of fusions remaining without Overlap in Healthy Brain GTEx Tissue: 30\n",
      "Number of fusions removed: 5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 1 ID</th>\n",
       "      <th>Gene 1 Strand</th>\n",
       "      <th>Gene 1 Reference Strand</th>\n",
       "      <th>Gene 1 Breakpoint</th>\n",
       "      <th>Gene 1 Input Position</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Gene 2 ID</th>\n",
       "      <th>Gene 2 Strand</th>\n",
       "      <th>Gene 2 Reference Strand</th>\n",
       "      <th>Gene 2 Breakpoint</th>\n",
       "      <th>Gene 2 Input Position</th>\n",
       "      <th>Breakpoint Distance</th>\n",
       "      <th>Supporting Reads</th>\n",
       "      <th>Read IDs</th>\n",
       "      <th>GF Program</th>\n",
       "      <th>Strand Comparison</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Gene 1 Gene Type</th>\n",
       "      <th>Gene 2 Gene Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290410</td>\n",
       "      <td>588-2281,39-1762,20-1940,137-389,30-1467,28-17...</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>25-587,1763-3510,1941-3680,22-136,1472-2045,17...</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>40</td>\n",
       "      <td>13581df7-a34b-4968-a43d-f984d302e89c</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NUP214:XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+,+,-,-,-,-,+,+,-,-</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199013</td>\n",
       "      <td>31-118,32-1068,1044-2974,1141-1269,1165-2437,2...</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-,-,+,+,+,+,-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>116-2661,1066-3601,57-1046,42-1143,62-1167,48-...</td>\n",
       "      <td>inf</td>\n",
       "      <td>10</td>\n",
       "      <td>002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XKR3:NUP214</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>9</td>\n",
       "      <td>e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Fusion Name Gene 1 Symbol           Gene 1 ID                                      Gene 1 Strand Gene 1 Reference Strand Gene 1 Breakpoint                              Gene 1 Input Position Gene 2 Symbol           Gene 2 ID                                      Gene 2 Strand Gene 2 Reference Strand Gene 2 Breakpoint                              Gene 2 Input Position  Breakpoint Distance  Supporting Reads                                           Read IDs    GF Program Strand Comparison       Sample Name Gene 1 Gene Type Gene 2 Gene Type\n",
       "1     BCR:ABL1           BCR  ENSG00000186716.21  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr22:23290410  588-2281,39-1762,20-1940,137-389,30-1467,28-17...          ABL1  ENSG00000097007.20  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr9:130854063  25-587,1763-3510,1941-3680,22-136,1472-2045,17...                  inf                52  0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "2     BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290412                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854063                                                NaN                  inf                52  f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "3     BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854064                                                NaN                  inf                40               13581df7-a34b-4968-a43d-f984d302e89c        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "6  NUP214:XKR3        NUP214  ENSG00000126883.19                                +,+,-,-,-,-,+,+,-,-                       +    chr9:131199013  31-118,32-1068,1044-2974,1141-1269,1165-2437,2...          XKR3   ENSG00000172967.8                                -,-,+,+,+,+,-,-,+,+                       -    chr22:16808085  116-2661,1066-3601,57-1046,42-1143,62-1167,48-...                  inf                10  002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding\n",
       "7  XKR3:NUP214          XKR3   ENSG00000172967.8                                                  -                       -    chr22:16808085                                                NaN        NUP214  ENSG00000126883.19                                                  +                       +    chr9:131199011                                                NaN                  inf                 9  e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "################################\n",
    "# GTEx Brain Tissue GF Overlap #\n",
    "################################\n",
    "\n",
    "gtex_brain_gf_df = pd.read_csv(gtex_brain_gfs_tsv, sep='\\t')\n",
    "\n",
    "# remove entries in gtex_brain_gf_df with matching Fusion Name, Gene 1 Symbol, and Gene 2 Symbol from no_ig_count\n",
    "# Merge the two dataframes on the relevant columns and find the matching rows\n",
    "merged = filtered_gene_types_protein_coding.merge(gtex_brain_gf_df[['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol']], \n",
    "                           on=['Gene 1 Symbol', 'Gene 2 Symbol'], \n",
    "                           how='inner')\n",
    "\n",
    "# Find reversed pairs (Gene 1 and Gene 2 swapped)\n",
    "reversed_merged = filtered_gene_types_protein_coding.merge(gtex_brain_gf_df[['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol']], \n",
    "                                    left_on=['Gene 1 Symbol', 'Gene 2 Symbol'], \n",
    "                                    right_on=['Gene 2 Symbol', 'Gene 1 Symbol'], \n",
    "                                    how='inner')\n",
    "\n",
    "# Combine both the merged and reversed entries\n",
    "combined_merged = pd.concat([merged, reversed_merged])\n",
    "\n",
    "# Remove any duplicates that may have been introduced by the merge\n",
    "combined_merged = combined_merged.drop_duplicates()\n",
    "\n",
    "\n",
    "# Now, remove the rows that match the combined merged DataFrame\n",
    "removed_gtex_brain_gfs = consistent_filter_df[~consistent_filter_df[['Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, 1).isin(\n",
    "    combined_merged[['Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, 1))]\n",
    "\n",
    "# Print or work with the resulting dataframe\n",
    "num_removed_gtex = consistent_filter_df.shape[0] - removed_gtex_brain_gfs.shape[0]\n",
    "\n",
    "\n",
    "# Display the results\n",
    "print('Number of fusions remaining without Overlap in Healthy Brain GTEx Tissue:', removed_gtex_brain_gfs.shape[0])\n",
    "print(\"Number of fusions removed:\", num_removed_gtex)\n",
    "removed_gtex_brain_gfs.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "36067c64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Initial Detected Unique GFs</th>\n",
       "      <th>Filter on CHOP and Cancer Census Genes</th>\n",
       "      <th>Filter on Read Support</th>\n",
       "      <th>Filter on Unrelated Genes</th>\n",
       "      <th>Filter on Consistent Strandness</th>\n",
       "      <th>Filter on GFs in GTEx Healthy Brain Samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>4031</td>\n",
       "      <td>468</td>\n",
       "      <td>26</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Initial Detected Unique GFs  Filter on CHOP and Cancer Census Genes  Filter on Read Support  Filter on Unrelated Genes  Filter on Consistent Strandness  Filter on GFs in GTEx Healthy Brain Samples\n",
       "0  sampleK562_pool2                         4031                                     468                      26                         18                               17                                           16"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_gtex_filter5 = removed_gtex_brain_gfs.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_gtex_filter5, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Filter on GFs in GTEx Healthy Brain Samples'}, inplace=True)\n",
    "fusion_counts_while_filtering['Filter on GFs in GTEx Healthy Brain Samples'] = fusion_counts_while_filtering['Filter on GFs in GTEx Healthy Brain Samples'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75db669f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9088703e",
   "metadata": {},
   "source": [
    "# 4. Fusion Annotation\n",
    "## **a.** Cross-Referencing Detected Fusions with External Databases \n",
    "- Mitelman, COSMIC, and ChimerDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b098f4b1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Literature Cross Reference \n",
    "\n",
    "# Mitelman DB\n",
    "mitelman_ref = pd.read_csv(mitelman_db_gfs, delimiter='\\t', header=None)\n",
    "mitelman_ref.columns = ['Fusion Name']\n",
    "mitelman_ref['Fusion Name'] = mitelman_ref['Fusion Name'].str.strip()\n",
    "\n",
    "split_fusions = mitelman_ref['Fusion Name'].str.split(':', n=1)  \n",
    "mitelman_ref[['Gene 1', 'Gene 2']] = pd.DataFrame(split_fusions.tolist(), index=mitelman_ref.index)\n",
    "mitelman_ref = mitelman_ref.drop(columns=['Fusion Name'])\n",
    "mitelman_ref.drop_duplicates()\n",
    "\n",
    "# Chimer DB\n",
    "chimer_ref = pd.read_csv(chimer_db_gfs, delimiter='\\t', header=None)\n",
    "chimer_ref.columns = ['Fusion Name']\n",
    "chimer_ref['Fusion Name'] = chimer_ref['Fusion Name'].str.strip()\n",
    "\n",
    "split_fusions = chimer_ref['Fusion Name'].str.split(':', n=1)  \n",
    "chimer_ref[['Gene 1', 'Gene 2']] = pd.DataFrame(split_fusions.tolist(), index=chimer_ref.index)\n",
    "chimer_ref = chimer_ref.drop(columns=['Fusion Name'])\n",
    "chimer_ref.drop_duplicates()\n",
    "\n",
    "# Cosmic DB\n",
    "cosmic_ref = pd.read_csv(cosmic_db_gfs, delimiter='\\t', header=None)\n",
    "cosmic_ref.columns = ['Fusion Name']\n",
    "cosmic_ref['Fusion Name'] = cosmic_ref['Fusion Name'].str.strip()\n",
    "\n",
    "split_fusions = cosmic_ref['Fusion Name'].str.split(':', n=1)  \n",
    "cosmic_ref[['Gene 1', 'Gene 2']] = pd.DataFrame(split_fusions.tolist(), index=cosmic_ref.index)\n",
    "cosmic_ref = cosmic_ref.drop(columns=['Fusion Name'])\n",
    "cosmic_ref.drop_duplicates()\n",
    "\n",
    "# All Fusion References\n",
    "all_fusions = pd.concat([mitelman_ref, chimer_ref, cosmic_ref])\n",
    "unique_fusions = all_fusions.drop_duplicates(subset=['Gene 1', 'Gene 2'])\n",
    "\n",
    "# Convert the reference files into sets for faster lookup\n",
    "mitelman_set = set(mitelman_ref.apply(lambda row: f\"{row['Gene 1']}:{row['Gene 2']}\", axis=1).dropna())\n",
    "chimer_set = set(chimer_ref.apply(lambda row: f\"{row['Gene 1']}:{row['Gene 2']}\", axis=1).dropna())\n",
    "cosmic_set = set(cosmic_ref.apply(lambda row: f\"{row['Gene 1']}:{row['Gene 2']}\", axis=1).dropna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a24286c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total unique fusions: 3 in 1 samples.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Number of Unique Known GFs</th>\n",
       "      <th>Unique Gene Fusions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>3</td>\n",
       "      <td>BCR:ABL1, NUP214:XKR3, XKR3:NUP214</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Number of Unique Known GFs                 Unique Gene Fusions\n",
       "0  sampleK562_pool2                           3  BCR:ABL1, NUP214:XKR3, XKR3:NUP214"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 1 ID</th>\n",
       "      <th>Gene 1 Strand</th>\n",
       "      <th>Gene 1 Reference Strand</th>\n",
       "      <th>Gene 1 Breakpoint</th>\n",
       "      <th>Gene 1 Input Position</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Gene 2 ID</th>\n",
       "      <th>Gene 2 Strand</th>\n",
       "      <th>Gene 2 Reference Strand</th>\n",
       "      <th>Gene 2 Breakpoint</th>\n",
       "      <th>Gene 2 Input Position</th>\n",
       "      <th>Breakpoint Distance</th>\n",
       "      <th>Supporting Reads</th>\n",
       "      <th>Read IDs</th>\n",
       "      <th>GF Program</th>\n",
       "      <th>Strand Comparison</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Gene 1 Gene Type</th>\n",
       "      <th>Gene 2 Gene Type</th>\n",
       "      <th>Literature Cross Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290410</td>\n",
       "      <td>588-2281,39-1762,20-1940,137-389,30-1467,28-17...</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>25-587,1763-3510,1941-3680,22-136,1472-2045,17...</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>40</td>\n",
       "      <td>13581df7-a34b-4968-a43d-f984d302e89c</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NUP214:XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+,+,-,-,-,-,+,+,-,-</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199013</td>\n",
       "      <td>31-118,32-1068,1044-2974,1141-1269,1165-2437,2...</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-,-,+,+,+,+,-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>116-2661,1066-3601,57-1046,42-1143,62-1167,48-...</td>\n",
       "      <td>inf</td>\n",
       "      <td>10</td>\n",
       "      <td>002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XKR3:NUP214</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>9</td>\n",
       "      <td>e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NUP214:XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808083</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>9</td>\n",
       "      <td>24016df2-c090-4de4-8fff-80c7d29648c9</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130780368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>7</td>\n",
       "      <td>c17e7c8e-0149-4d8b-9237-26b5858ff694,03d7adcb-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Fusion Name Gene 1 Symbol           Gene 1 ID                                      Gene 1 Strand Gene 1 Reference Strand Gene 1 Breakpoint                              Gene 1 Input Position Gene 2 Symbol           Gene 2 ID                                      Gene 2 Strand Gene 2 Reference Strand Gene 2 Breakpoint                              Gene 2 Input Position  Breakpoint Distance  Supporting Reads                                           Read IDs    GF Program Strand Comparison       Sample Name Gene 1 Gene Type Gene 2 Gene Type                         Literature Cross Reference\n",
       "1      BCR:ABL1           BCR  ENSG00000186716.21  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr22:23290410  588-2281,39-1762,20-1940,137-389,30-1467,28-17...          ABL1  ENSG00000097007.20  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr9:130854063  25-587,1763-3510,1941-3680,22-136,1472-2045,17...                  inf                52  0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R...\n",
       "2      BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290412                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854063                                                NaN                  inf                52  f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R...\n",
       "3      BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854064                                                NaN                  inf                40               13581df7-a34b-4968-a43d-f984d302e89c        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R...\n",
       "6   NUP214:XKR3        NUP214  ENSG00000126883.19                                +,+,-,-,-,-,+,+,-,-                       +    chr9:131199013  31-118,32-1068,1044-2974,1141-1269,1165-2437,2...          XKR3   ENSG00000172967.8                                -,-,+,+,+,+,-,-,+,+                       -    chr22:16808085  116-2661,1066-3601,57-1046,42-1143,62-1167,48-...                  inf                10  002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding                                 Mitelman Reference\n",
       "7   XKR3:NUP214          XKR3   ENSG00000172967.8                                                  -                       -    chr22:16808085                                                NaN        NUP214  ENSG00000126883.19                                                  +                       +    chr9:131199011                                                NaN                  inf                 9  e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding                                 Mitelman Reference\n",
       "9   NUP214:XKR3        NUP214  ENSG00000126883.19                                                  +                       +    chr9:131199015                                                NaN          XKR3   ENSG00000172967.8                                                  -                       -    chr22:16808083                                                NaN                  inf                 9               24016df2-c090-4de4-8fff-80c7d29648c9        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding                                 Mitelman Reference\n",
       "12     BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130780368                                                NaN                  inf                 7  c17e7c8e-0149-4d8b-9237-26b5858ff694,03d7adcb-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create a function to check if fusion is present in any reference file\n",
    "def check_fusion_in_references(row):\n",
    "    fusion1 = f\"{row['Gene 1 Symbol']}:{row['Gene 2 Symbol']}\"\n",
    "    fusion2 = f\"{row['Gene 2 Symbol']}:{row['Gene 1 Symbol']}\"  # Reverse orientation\n",
    "    result = []\n",
    "\n",
    "    # Check against each reference set\n",
    "    if fusion1 in mitelman_set or fusion2 in mitelman_set:\n",
    "        result.append('Mitelman Reference')\n",
    "    if fusion1 in chimer_set or fusion2 in chimer_set:\n",
    "        result.append('Chimer Reference')\n",
    "    if fusion1 in cosmic_set or fusion2 in cosmic_set:\n",
    "        result.append('COSMIC Reference')\n",
    "\n",
    "    # Return the result as a comma-separated string\n",
    "    return ', '.join(result) if result else 'No match'\n",
    "\n",
    "# Apply the function to each row of the dataframe\n",
    "filter_df_literature = removed_gtex_brain_gfs.copy()\n",
    "filter_df_literature['Literature Cross Reference'] = filter_df_literature.apply(check_fusion_in_references, axis=1)\n",
    "known_gfs_detected_df = filter_df_literature[filter_df_literature['Literature Cross Reference'] != 'No match']\n",
    "\n",
    "\n",
    "# Show overview of known fusions by sample \n",
    "unique_gfs_per_sample = (\n",
    "    known_gfs_detected_df.groupby('Sample Name')['Fusion Name']\n",
    "    .apply(lambda x: \", \".join(sorted(set(x))))  \n",
    "    .reset_index(name='Unique Gene Fusions'))\n",
    "\n",
    "# Add count of unique fusions per sample\n",
    "unique_gfs_per_sample.insert(1, 'Number of Unique Known GFs', unique_gfs_per_sample['Unique Gene Fusions'].apply(lambda x: x.count(\",\") + 1 if x else 0))\n",
    "unique_gfs_per_sample.sort_values(by='Sample Name', inplace=True)\n",
    "\n",
    "\n",
    "# Get the total number of unique fusions\n",
    "num_unique_fusions = len(known_gfs_detected_df['Fusion Name'].value_counts())\n",
    "num_unique_samples = len(known_gfs_detected_df['Sample Name'].value_counts())\n",
    "print(f\"Total unique fusions: {num_unique_fusions} in {num_unique_samples} samples.\")\n",
    "display(unique_gfs_per_sample)\n",
    "\n",
    "\n",
    "# look at those reported in literature only .. just annotation no filtering!\n",
    "display(known_gfs_detected_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "83b1f3f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Initial Detected Unique GFs</th>\n",
       "      <th>Filter on CHOP and Cancer Census Genes</th>\n",
       "      <th>Filter on Read Support</th>\n",
       "      <th>Filter on Unrelated Genes</th>\n",
       "      <th>Filter on Consistent Strandness</th>\n",
       "      <th>Filter on GFs in GTEx Healthy Brain Samples</th>\n",
       "      <th>GFs in Fusion Reference Databases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>4031</td>\n",
       "      <td>468</td>\n",
       "      <td>26</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Initial Detected Unique GFs  Filter on CHOP and Cancer Census Genes  Filter on Read Support  Filter on Unrelated Genes  Filter on Consistent Strandness  Filter on GFs in GTEx Healthy Brain Samples  GFs in Fusion Reference Databases\n",
       "0  sampleK562_pool2                         4031                                     468                      26                         18                               17                                           16                                  3"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_known_ref_filter6 = known_gfs_detected_df.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_known_ref_filter6, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'GFs in Fusion Reference Databases'}, inplace=True)\n",
    "fusion_counts_while_filtering['GFs in Fusion Reference Databases'] = fusion_counts_while_filtering['GFs in Fusion Reference Databases'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ae0a597",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e3eefdfe",
   "metadata": {},
   "source": [
    "##  **b.** Remove Recurrent Fusions in More Than a Certain Percentage of the Samples  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "81a88055",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Recurrent Fusions Removed Filter: \n",
      "Total samples: 1, using a 15.0% threshold: 0.15 occurrences within a sample\n",
      "Recurrent fusions\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Sample Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATIC:MYL6</td>\n",
       "      <td>ATIC</td>\n",
       "      <td>MYL6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATP5PD:GATA2</td>\n",
       "      <td>ATP5PD</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GGPS1:PAFAH1B2</td>\n",
       "      <td>GGPS1</td>\n",
       "      <td>PAFAH1B2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KCTD2:GATA2</td>\n",
       "      <td>KCTD2</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MALAT1:AP2S1</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MAX:PABPC1</td>\n",
       "      <td>MAX</td>\n",
       "      <td>PABPC1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MYL6:ATIC</td>\n",
       "      <td>MYL6</td>\n",
       "      <td>ATIC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NACA:TBL1XR1</td>\n",
       "      <td>NACA</td>\n",
       "      <td>TBL1XR1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NF2:C9orf78</td>\n",
       "      <td>NF2</td>\n",
       "      <td>C9orf78</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NUP214:XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PAFAH1B2:GGPS1</td>\n",
       "      <td>PAFAH1B2</td>\n",
       "      <td>GGPS1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SBNO1:TFRC</td>\n",
       "      <td>SBNO1</td>\n",
       "      <td>TFRC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TFRC:SBNO1</td>\n",
       "      <td>TFRC</td>\n",
       "      <td>SBNO1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TMEM9:MAPK1</td>\n",
       "      <td>TMEM9</td>\n",
       "      <td>MAPK1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>XKR3:NUP214</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Fusion Name Gene 1 Symbol Gene 2 Symbol  Sample Count\n",
       "0        ATIC:MYL6          ATIC          MYL6             1\n",
       "1     ATP5PD:GATA2        ATP5PD         GATA2             1\n",
       "2         BCR:ABL1           BCR          ABL1             1\n",
       "3   GGPS1:PAFAH1B2         GGPS1      PAFAH1B2             1\n",
       "4      KCTD2:GATA2         KCTD2         GATA2             1\n",
       "5     MALAT1:AP2S1        MALAT1         AP2S1             1\n",
       "6       MAX:PABPC1           MAX        PABPC1             1\n",
       "7        MYL6:ATIC          MYL6          ATIC             1\n",
       "8     NACA:TBL1XR1          NACA       TBL1XR1             1\n",
       "9      NF2:C9orf78           NF2       C9orf78             1\n",
       "10     NUP214:XKR3        NUP214          XKR3             1\n",
       "11  PAFAH1B2:GGPS1      PAFAH1B2         GGPS1             1\n",
       "12      SBNO1:TFRC         SBNO1          TFRC             1\n",
       "13      TFRC:SBNO1          TFRC         SBNO1             1\n",
       "14     TMEM9:MAPK1         TMEM9         MAPK1             1\n",
       "15     XKR3:NUP214          XKR3        NUP214             1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original entries matching the recurrent fusions: 30\n",
      "\n",
      "Number of rows removed: 23\n",
      "Recurrent GFs Removed, Remaining GFs: (7, 22)\n",
      "\n",
      "Recurrent fusions found in literature and not removed:\n",
      "BCR:ABL1\n",
      "NUP214:XKR3\n",
      "XKR3:NUP214\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 1 ID</th>\n",
       "      <th>Gene 1 Strand</th>\n",
       "      <th>Gene 1 Reference Strand</th>\n",
       "      <th>Gene 1 Breakpoint</th>\n",
       "      <th>Gene 1 Input Position</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Gene 2 ID</th>\n",
       "      <th>Gene 2 Strand</th>\n",
       "      <th>Gene 2 Reference Strand</th>\n",
       "      <th>Gene 2 Breakpoint</th>\n",
       "      <th>Gene 2 Input Position</th>\n",
       "      <th>Breakpoint Distance</th>\n",
       "      <th>Supporting Reads</th>\n",
       "      <th>Read IDs</th>\n",
       "      <th>GF Program</th>\n",
       "      <th>Strand Comparison</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Gene 1 Gene Type</th>\n",
       "      <th>Gene 2 Gene Type</th>\n",
       "      <th>Literature Cross Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290410</td>\n",
       "      <td>588-2281,39-1762,20-1940,137-389,30-1467,28-17...</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>25-587,1763-3510,1941-3680,22-136,1472-2045,17...</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>40</td>\n",
       "      <td>13581df7-a34b-4968-a43d-f984d302e89c</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NUP214:XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+,+,-,-,-,-,+,+,-,-</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199013</td>\n",
       "      <td>31-118,32-1068,1044-2974,1141-1269,1165-2437,2...</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-,-,+,+,+,+,-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>116-2661,1066-3601,57-1046,42-1143,62-1167,48-...</td>\n",
       "      <td>inf</td>\n",
       "      <td>10</td>\n",
       "      <td>002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XKR3:NUP214</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>9</td>\n",
       "      <td>e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Fusion Name Gene 1 Symbol           Gene 1 ID                                      Gene 1 Strand Gene 1 Reference Strand Gene 1 Breakpoint                              Gene 1 Input Position Gene 2 Symbol           Gene 2 ID                                      Gene 2 Strand Gene 2 Reference Strand Gene 2 Breakpoint                              Gene 2 Input Position  Breakpoint Distance  Supporting Reads                                           Read IDs    GF Program Strand Comparison       Sample Name Gene 1 Gene Type Gene 2 Gene Type                         Literature Cross Reference\n",
       "1     BCR:ABL1           BCR  ENSG00000186716.21  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr22:23290410  588-2281,39-1762,20-1940,137-389,30-1467,28-17...          ABL1  ENSG00000097007.20  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr9:130854063  25-587,1763-3510,1941-3680,22-136,1472-2045,17...                  inf                52  0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R...\n",
       "2     BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290412                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854063                                                NaN                  inf                52  f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R...\n",
       "3     BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854064                                                NaN                  inf                40               13581df7-a34b-4968-a43d-f984d302e89c        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R...\n",
       "6  NUP214:XKR3        NUP214  ENSG00000126883.19                                +,+,-,-,-,-,+,+,-,-                       +    chr9:131199013  31-118,32-1068,1044-2974,1141-1269,1165-2437,2...          XKR3   ENSG00000172967.8                                -,-,+,+,+,+,-,-,+,+                       -    chr22:16808085  116-2661,1066-3601,57-1046,42-1143,62-1167,48-...                  inf                10  002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding                                 Mitelman Reference\n",
       "7  XKR3:NUP214          XKR3   ENSG00000172967.8                                                  -                       -    chr22:16808085                                                NaN        NUP214  ENSG00000126883.19                                                  +                       +    chr9:131199011                                                NaN                  inf                 9  e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding                                 Mitelman Reference"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Recurrent Fusions Removed Filter: \")\n",
    "\n",
    "# Count unique samples each fusion appears in\n",
    "fusion_sample_counts = filter_df_literature.groupby(['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol'])['Sample Name'].nunique()\n",
    "\n",
    "# Calculate threshold and filter based on occurrence\n",
    "total_samples = filter_df_literature['Sample Name'].nunique()\n",
    "\n",
    "threshold = total_samples * percentage_samples\n",
    "\n",
    "recurrent_fusions = fusion_sample_counts[fusion_sample_counts > threshold]\n",
    "recurrent_fusions_df = recurrent_fusions.reset_index()\n",
    "recurrent_fusions_df.rename(columns={'Sample Name': 'Sample Count'}, inplace=True)\n",
    "\n",
    "# Output\n",
    "print(f\"Total samples: {total_samples}, using a {percentage_samples*100}% threshold: {threshold} occurrences within a sample\")\n",
    "print(\"Recurrent fusions\")\n",
    "display(recurrent_fusions_df)\n",
    "\n",
    "# Show original entries \n",
    "original_recurrent_entries_df = filter_df_literature[\n",
    "    filter_df_literature[['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol']].apply(\n",
    "        tuple, axis=1\n",
    "    ).isin(recurrent_fusions_df[['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, axis=1))\n",
    "]\n",
    "print(f\"Original entries matching the recurrent fusions: {original_recurrent_entries_df.shape[0]}\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### EXCLUDE: Fusions Identified in Online Databases from Recurrent Removal ###\n",
    "\n",
    "# Identify fusions in literature (direct match or reciprocal)\n",
    "if 'Literature Cross Reference' in filter_df_literature.columns:\n",
    "    literature_fusions = set(\n",
    "        filter_df_literature[\n",
    "            filter_df_literature['Literature Cross Reference'] != 'No match'\n",
    "        ][['Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, axis=1)\n",
    "    )\n",
    "else:\n",
    "    literature_fusions = set()\n",
    "\n",
    "if 'Literature Reciprocal Fusion' in filter_df_literature.columns:\n",
    "    reciprocal_literature_fusions = set(\n",
    "        filter_df_literature[\n",
    "            filter_df_literature['Literature Reciprocal Fusion'] == 'Yes'\n",
    "        ][['Gene 1 Symbol', 'Gene 2 Symbol']].apply(lambda x: (x[1], x[0]), axis=1)\n",
    "    )\n",
    "else:\n",
    "    reciprocal_literature_fusions = set()\n",
    "\n",
    "# Combine both sets\n",
    "literature_fusions |= reciprocal_literature_fusions\n",
    "\n",
    "# Create a set of recurrent fusion pairs (both orientations)\n",
    "recurrent_fusion_pairs = set(\n",
    "    list(zip(recurrent_fusions.index.get_level_values('Gene 1 Symbol'), recurrent_fusions.index.get_level_values('Gene 2 Symbol'))) +\n",
    "    list(zip(recurrent_fusions.index.get_level_values('Gene 2 Symbol'), recurrent_fusions.index.get_level_values('Gene 1 Symbol')))\n",
    ")\n",
    "\n",
    "# Exclude fusions in literature from removal\n",
    "filtered_fusion_pairs = recurrent_fusion_pairs - literature_fusions\n",
    "\n",
    "# Filter out only the **non-literature** recurrent fusions\n",
    "recurrent_filter_df = filter_df_literature[\n",
    "    ~filter_df_literature[['Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, axis=1).isin(filtered_fusion_pairs)\n",
    "]\n",
    "\n",
    "# Results\n",
    "rows_removed = filter_df_literature.shape[0] - recurrent_filter_df.shape[0]\n",
    "print(f\"\\nNumber of rows removed: {rows_removed}\")\n",
    "print(f\"Recurrent GFs Removed, Remaining GFs: {recurrent_filter_df.shape}\")\n",
    "\n",
    "# Show which recurrent fusions were found in literature and not removed\n",
    "literature_recurrent_fusions = recurrent_fusion_pairs & literature_fusions\n",
    "literature_recurrent_fusions_df = filter_df_literature[\n",
    "    filter_df_literature[['Gene 1 Symbol', 'Gene 2 Symbol']].apply(tuple, axis=1).isin(literature_recurrent_fusions)\n",
    "]\n",
    "\n",
    "print(\"\\nRecurrent fusions found in literature and not removed:\")\n",
    "print(\"\\n\".join(literature_recurrent_fusions_df['Fusion Name'].unique()))\n",
    "\n",
    "recurrent_filter_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "74d4d3f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Initial Detected Unique GFs</th>\n",
       "      <th>Filter on CHOP and Cancer Census Genes</th>\n",
       "      <th>Filter on Read Support</th>\n",
       "      <th>Filter on Unrelated Genes</th>\n",
       "      <th>Filter on Consistent Strandness</th>\n",
       "      <th>Filter on GFs in GTEx Healthy Brain Samples</th>\n",
       "      <th>GFs in Fusion Reference Databases</th>\n",
       "      <th>Filter on Recurrent in 15.0% of Samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>4031</td>\n",
       "      <td>468</td>\n",
       "      <td>26</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Initial Detected Unique GFs  Filter on CHOP and Cancer Census Genes  Filter on Read Support  Filter on Unrelated Genes  Filter on Consistent Strandness  Filter on GFs in GTEx Healthy Brain Samples  GFs in Fusion Reference Databases  Filter on Recurrent in 15.0% of Samples\n",
       "0  sampleK562_pool2                         4031                                     468                      26                         18                               17                                           16                                  3                                        3"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add counts to keep track of fusion counts while filtering\n",
    "fusion_counts_recurrent_GF_filter8 = recurrent_filter_df.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_recurrent_GF_filter8, on='Sample Name', how='left')\n",
    "\n",
    "recurrent_col_name = f'Filter on Recurrent in {percentage_samples*100}% of Samples'\n",
    "\n",
    "fusion_counts_while_filtering.rename(\n",
    "    columns={'Fusion Name': recurrent_col_name}, \n",
    "    inplace=True)\n",
    "\n",
    "# Fill NaN values and convert to int\n",
    "fusion_counts_while_filtering[recurrent_col_name] = fusion_counts_while_filtering[recurrent_col_name].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db691afe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6d261d19",
   "metadata": {},
   "source": [
    "## **c.** CNS and Tumor Morphology Prioritization Using Mitelman DB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c7f98398",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Genes with Associated Diseases: 2779 \n",
      "\n",
      "Number of entries with at least one brain disease association: 7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fusion Name</th>\n",
       "      <th>Gene 1 Symbol</th>\n",
       "      <th>Gene 1 ID</th>\n",
       "      <th>Gene 1 Strand</th>\n",
       "      <th>Gene 1 Reference Strand</th>\n",
       "      <th>Gene 1 Breakpoint</th>\n",
       "      <th>Gene 1 Input Position</th>\n",
       "      <th>Gene 2 Symbol</th>\n",
       "      <th>Gene 2 ID</th>\n",
       "      <th>Gene 2 Strand</th>\n",
       "      <th>Gene 2 Reference Strand</th>\n",
       "      <th>Gene 2 Breakpoint</th>\n",
       "      <th>Gene 2 Input Position</th>\n",
       "      <th>Breakpoint Distance</th>\n",
       "      <th>Supporting Reads</th>\n",
       "      <th>Read IDs</th>\n",
       "      <th>GF Program</th>\n",
       "      <th>Strand Comparison</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Gene 1 Gene Type</th>\n",
       "      <th>Gene 2 Gene Type</th>\n",
       "      <th>Literature Cross Reference</th>\n",
       "      <th>Gene 1 Brain Disease Associated</th>\n",
       "      <th>Gene 2 Brain Disease Associated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290410</td>\n",
       "      <td>588-2281,39-1762,20-1940,137-389,30-1467,28-17...</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>-,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>25-587,1763-3510,1941-3680,22-136,1472-2045,17...</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "      <td>Astrocytoma, grade I-II, Glioma, NOS, Astrocyt...</td>\n",
       "      <td>Astrocytoma, grade III-IV/Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>52</td>\n",
       "      <td>f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "      <td>Astrocytoma, grade I-II, Glioma, NOS, Astrocyt...</td>\n",
       "      <td>Astrocytoma, grade III-IV/Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BCR:ABL1</td>\n",
       "      <td>BCR</td>\n",
       "      <td>ENSG00000186716.21</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22:23290413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007.20</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:130854064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>40</td>\n",
       "      <td>13581df7-a34b-4968-a43d-f984d302e89c</td>\n",
       "      <td>JAFFAL</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference, Chimer Reference, COSMIC R...</td>\n",
       "      <td>Astrocytoma, grade I-II, Glioma, NOS, Astrocyt...</td>\n",
       "      <td>Astrocytoma, grade III-IV/Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NUP214:XKR3</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+,+,-,-,-,-,+,+,-,-</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199013</td>\n",
       "      <td>31-118,32-1068,1044-2974,1141-1269,1165-2437,2...</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-,-,+,+,+,+,-,-,+,+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>116-2661,1066-3601,57-1046,42-1143,62-1167,48-...</td>\n",
       "      <td>inf</td>\n",
       "      <td>10</td>\n",
       "      <td>002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...</td>\n",
       "      <td>LongGF</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference</td>\n",
       "      <td>Astrocytoma, grade I-II</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XKR3:NUP214</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>ENSG00000172967.8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22:16808085</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NUP214</td>\n",
       "      <td>ENSG00000126883.19</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9:131199011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inf</td>\n",
       "      <td>9</td>\n",
       "      <td>e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...</td>\n",
       "      <td>FusionSeeker</td>\n",
       "      <td>Consistent</td>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mitelman Reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Astrocytoma, grade I-II</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Fusion Name Gene 1 Symbol           Gene 1 ID                                      Gene 1 Strand Gene 1 Reference Strand Gene 1 Breakpoint                              Gene 1 Input Position Gene 2 Symbol           Gene 2 ID                                      Gene 2 Strand Gene 2 Reference Strand Gene 2 Breakpoint                              Gene 2 Input Position  Breakpoint Distance  Supporting Reads                                           Read IDs    GF Program Strand Comparison       Sample Name Gene 1 Gene Type Gene 2 Gene Type                         Literature Cross Reference                    Gene 1 Brain Disease Associated         Gene 2 Brain Disease Associated\n",
       "1     BCR:ABL1           BCR  ENSG00000186716.21  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr22:23290410  588-2281,39-1762,20-1940,137-389,30-1467,28-17...          ABL1  ENSG00000097007.20  -,+,+,-,+,+,-,-,-,+,-,-,-,+,+,+,+,-,+,-,+,+,-,...                       +    chr9:130854063  25-587,1763-3510,1941-3680,22-136,1472-2045,17...                  inf                52  0328aaf6-1b42-42c9-b505-d3b84df779f8,13581df7-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R...  Astrocytoma, grade I-II, Glioma, NOS, Astrocyt...  Astrocytoma, grade III-IV/Glioblastoma\n",
       "2     BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290412                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854063                                                NaN                  inf                52  f52eb887-3f94-4a20-8546-19f8695d180e,0328aaf6-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R...  Astrocytoma, grade I-II, Glioma, NOS, Astrocyt...  Astrocytoma, grade III-IV/Glioblastoma\n",
       "3     BCR:ABL1           BCR  ENSG00000186716.21                                                  +                       +    chr22:23290413                                                NaN          ABL1  ENSG00000097007.20                                                  +                       +    chr9:130854064                                                NaN                  inf                40               13581df7-a34b-4968-a43d-f984d302e89c        JAFFAL        Consistent  sampleK562_pool2   protein_coding   protein_coding  Mitelman Reference, Chimer Reference, COSMIC R...  Astrocytoma, grade I-II, Glioma, NOS, Astrocyt...  Astrocytoma, grade III-IV/Glioblastoma\n",
       "6  NUP214:XKR3        NUP214  ENSG00000126883.19                                +,+,-,-,-,-,+,+,-,-                       +    chr9:131199013  31-118,32-1068,1044-2974,1141-1269,1165-2437,2...          XKR3   ENSG00000172967.8                                -,-,+,+,+,+,-,-,+,+                       -    chr22:16808085  116-2661,1066-3601,57-1046,42-1143,62-1167,48-...                  inf                10  002e5062-b79f-4d26-bb5a-f97073a7a635,24016df2-...        LongGF        Consistent  sampleK562_pool2   protein_coding   protein_coding                                 Mitelman Reference                            Astrocytoma, grade I-II                                     NaN\n",
       "7  XKR3:NUP214          XKR3   ENSG00000172967.8                                                  -                       -    chr22:16808085                                                NaN        NUP214  ENSG00000126883.19                                                  +                       +    chr9:131199011                                                NaN                  inf                 9  e0fe3f68-48e2-4487-be8a-85832a8c792e,24016df2-...  FusionSeeker        Consistent  sampleK562_pool2   protein_coding   protein_coding                                 Mitelman Reference                                                NaN                 Astrocytoma, grade I-II"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genes with cns related diseases GFs\n",
    "mitelman_brain_gfs_df = pd.read_csv(mitelman_brain_gfs_csv)\n",
    "\n",
    "# Dictionary to store gene-to-disease associations\n",
    "gene_disease_map = {}\n",
    "\n",
    "# Process the DataFrame to map genes to associated diseases\n",
    "for _, row in mitelman_brain_gfs_df.iterrows():\n",
    "    morph_name = row[\"MorphName\"]\n",
    "    # Split GeneShort by \"::\" and remove \"+\" characters\n",
    "    genes = row[\"GeneShort\"].replace(\"+\", \"\").split(\"::\")\n",
    "    for gene in genes:\n",
    "        # Append the disease to the existing list or start a new one\n",
    "        if gene in gene_disease_map:\n",
    "            if morph_name not in gene_disease_map[gene]:\n",
    "                gene_disease_map[gene].append(morph_name)\n",
    "        else:\n",
    "            gene_disease_map[gene] = [morph_name]\n",
    "\n",
    "# Convert the dictionary to a DataFrame\n",
    "result_rows = [{\"Gene Name\": gene, \"Brain Disease Associated\": \", \".join(diseases)} for gene, diseases in gene_disease_map.items()]\n",
    "mitelman_genes_diseases = pd.DataFrame(result_rows)\n",
    "print(\"Unique Genes with Associated Diseases:\", mitelman_genes_diseases.shape[0], '\\n')\n",
    "\n",
    "\n",
    "#####################\n",
    "\n",
    "# Create a dictionary from the disease association DataFrame for quick lookup\n",
    "gene_disease_map = dict(zip(mitelman_genes_diseases['Gene Name'], mitelman_genes_diseases['Brain Disease Associated']))\n",
    "\n",
    "#####################\n",
    "\n",
    "\n",
    "# Add a new column for Gene 1 Brain Disease Associated\n",
    "CNS_related_genes_df = recurrent_filter_df.copy()\n",
    "\n",
    "CNS_related_genes_df['Gene 1 Brain Disease Associated'] = CNS_related_genes_df['Gene 1 Symbol'].map(gene_disease_map)\n",
    "\n",
    "# Add a new column for Gene 2 Brain Disease Associated\n",
    "CNS_related_genes_df['Gene 2 Brain Disease Associated'] = CNS_related_genes_df['Gene 2 Symbol'].map(gene_disease_map)\n",
    "\n",
    "# Count rows where at least one of the columns is not None\n",
    "non_none_count = CNS_related_genes_df[\n",
    "    CNS_related_genes_df['Gene 1 Brain Disease Associated'].notna() | \n",
    "    CNS_related_genes_df['Gene 2 Brain Disease Associated'].notna()].shape[0]\n",
    "\n",
    "print(f\"Number of entries with at least one brain disease association: {non_none_count}\")\n",
    "CNS_related_genes_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dd9b4c3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entries where both 'Gene 1 Brain Disease Associated' and 'Gene 2 Brain Disease Associated' are not NaN: 1\n",
      "Entries where only one of 'Gene 1 Brain Disease Associated' or 'Gene 2 Brain Disease Associated' is NaN: 2\n",
      "Entries where both 'Gene 1 Brain Disease Associated' and 'Gene 2 Brain Disease Associated' are NaN: 0\n"
     ]
    }
   ],
   "source": [
    "# Drop duplicates based on 'Fusion Name' and 'Sample Name' columns first\n",
    "CNS_related_genes_df_count_brain_genes = CNS_related_genes_df.drop_duplicates(subset=['Fusion Name', 'Sample Name'])\n",
    "\n",
    "# Count entries where both 'Gene 1 Brain Disease Associated' and 'Gene 2 Brain Disease Associated' are not NaN\n",
    "both_not_nan = CNS_related_genes_df_count_brain_genes.dropna(subset=['Gene 1 Brain Disease Associated', 'Gene 2 Brain Disease Associated'])\n",
    "\n",
    "# Count entries where only one of 'Gene 1 Brain Disease Associated' or 'Gene 2 Brain Disease Associated' is NaN\n",
    "one_nan = CNS_related_genes_df_count_brain_genes[\n",
    "    CNS_related_genes_df_count_brain_genes['Gene 1 Brain Disease Associated'].isna() ^ \n",
    "    CNS_related_genes_df_count_brain_genes['Gene 2 Brain Disease Associated'].isna()\n",
    "]\n",
    "\n",
    "\n",
    "# Count entries where only one of 'Gene 1 Brain Disease Associated' or 'Gene 2 Brain Disease Associated' is NaN\n",
    "both_nan = CNS_related_genes_df_count_brain_genes[\n",
    "    CNS_related_genes_df_count_brain_genes['Gene 1 Brain Disease Associated'].isna() & \n",
    "    CNS_related_genes_df_count_brain_genes['Gene 2 Brain Disease Associated'].isna()\n",
    "]\n",
    "\n",
    "# Display counts\n",
    "print(f\"Entries where both 'Gene 1 Brain Disease Associated' and 'Gene 2 Brain Disease Associated' are not NaN: {both_not_nan.shape[0]}\")\n",
    "print(f\"Entries where only one of 'Gene 1 Brain Disease Associated' or 'Gene 2 Brain Disease Associated' is NaN: {one_nan.shape[0]}\")\n",
    "\n",
    "print(f\"Entries where both 'Gene 1 Brain Disease Associated' and 'Gene 2 Brain Disease Associated' are NaN: {both_nan.shape[0]}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5d0c433a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Initial Detected Unique GFs</th>\n",
       "      <th>Filter on CHOP and Cancer Census Genes</th>\n",
       "      <th>Filter on Read Support</th>\n",
       "      <th>Filter on Unrelated Genes</th>\n",
       "      <th>Filter on Consistent Strandness</th>\n",
       "      <th>Filter on GFs in GTEx Healthy Brain Samples</th>\n",
       "      <th>GFs in Fusion Reference Databases</th>\n",
       "      <th>Filter on Recurrent in 15.0% of Samples</th>\n",
       "      <th>Genes of GF in CNS Related Diseases - Both Genes Associated</th>\n",
       "      <th>Genes of GF in CNS Related Diseases - Only One Gene Associated</th>\n",
       "      <th>Genes of GF in CNS Related Diseases - No Gene Associated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleK562_pool2</td>\n",
       "      <td>4031</td>\n",
       "      <td>468</td>\n",
       "      <td>26</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Name  Initial Detected Unique GFs  Filter on CHOP and Cancer Census Genes  Filter on Read Support  Filter on Unrelated Genes  Filter on Consistent Strandness  Filter on GFs in GTEx Healthy Brain Samples  GFs in Fusion Reference Databases  Filter on Recurrent in 15.0% of Samples  Genes of GF in CNS Related Diseases - Both Genes Associated  Genes of GF in CNS Related Diseases - Only One Gene Associated  Genes of GF in CNS Related Diseases - No Gene Associated\n",
       "0  sampleK562_pool2                         4031                                     468                      26                         18                               17                                           16                                  3                                        3                                                  1                                                            2                                                               0       "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# both genes in CNS GFs list of genes \n",
    "fusion_counts_CNS_gene_disease_filter7_both = both_not_nan.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_CNS_gene_disease_filter7_both, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Genes of GF in CNS Related Diseases - Both Genes Associated'}, inplace=True)\n",
    "fusion_counts_while_filtering['Genes of GF in CNS Related Diseases - Both Genes Associated'] = fusion_counts_while_filtering['Genes of GF in CNS Related Diseases - Both Genes Associated'].fillna(0).astype(int)\n",
    "\n",
    "#####\n",
    "\n",
    "# only ONE gene in CNS GFs list of genes \n",
    "fusion_counts_CNS_gene_disease_filter7_one = one_nan.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_CNS_gene_disease_filter7_one, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Genes of GF in CNS Related Diseases - Only One Gene Associated'}, inplace=True)\n",
    "fusion_counts_while_filtering['Genes of GF in CNS Related Diseases - Only One Gene Associated'] = fusion_counts_while_filtering['Genes of GF in CNS Related Diseases - Only One Gene Associated'].fillna(0).astype(int)\n",
    "\n",
    "#####\n",
    "\n",
    "# NO ASSOCIATION between the genes: \n",
    "fusion_counts_CNS_gene_disease_filter7_none = both_nan.groupby('Sample Name')['Fusion Name'].nunique().reset_index()\n",
    "\n",
    "# merge \n",
    "fusion_counts_while_filtering = fusion_counts_while_filtering.merge(fusion_counts_CNS_gene_disease_filter7_none, on='Sample Name', how='left')\n",
    "fusion_counts_while_filtering.rename(columns={'Fusion Name': 'Genes of GF in CNS Related Diseases - No Gene Associated'}, inplace=True)\n",
    "fusion_counts_while_filtering['Genes of GF in CNS Related Diseases - No Gene Associated'] = fusion_counts_while_filtering['Genes of GF in CNS Related Diseases - No Gene Associated'].fillna(0).astype(int)\n",
    "\n",
    "fusion_counts_while_filtering\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b87331d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "26a11ac2",
   "metadata": {},
   "source": [
    "# 5. Results Summary\n",
    "## **a.** Summary of Results  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "19124166",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File saved to: /mnt/isilon/wang_lab/karly/project/chop_panel_ont_aim1/github_upload/example_transcriptome_fusions/filtered_fusion_calls/summary_filtered_output.tsv\n",
      "File saved to: /mnt/isilon/wang_lab/karly/project/chop_panel_ont_aim1/github_upload/example_transcriptome_fusions/filtered_fusion_calls/summary_filtered_output_counts.tsv\n"
     ]
    }
   ],
   "source": [
    "# Save the Filtered Fusion Entries\n",
    "output_file_path = f'{filtered_output_save_folder}summary_filtered_output.tsv'\n",
    "recurrent_filter_df.to_csv(output_file_path, sep='\\t', index=False)\n",
    "print(f\"File saved to: {output_file_path}\")\n",
    "\n",
    "\n",
    "# Save the Counts of Fusions after each Filtering Step \n",
    "# Specify the output file path\n",
    "output_file_path_counts = f'{filtered_output_save_folder}summary_filtered_output_counts.tsv'\n",
    "fusion_counts_while_filtering.to_csv(output_file_path_counts, sep='\\t', index=False)\n",
    "print(f\"File saved to: {output_file_path_counts}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a089ea4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results Summary\n",
      "--------------------------------------------------\n",
      "Total number of GF entries remaining: 3\n",
      "--------------------------------------------------\n",
      "Number of samples with no fusions: 0 []\n",
      "Number of samples with only one fusion: 0 []\n",
      "Number of samples with more than one fusion: 1 ['sampleK562_pool2']\n",
      "--------------------------------------------------\n",
      "Minimum number of fusions: 3\n",
      "Maximum number of fusions: 3\n",
      "Mean number of fusions: 3\n",
      "--------------------------------------------------\n",
      "GF Program Counts:\n",
      "GF Program\n",
      "LongGF          2\n",
      "FusionSeeker    1\n",
      "Name: count, dtype: int64\n",
      "--------------------------------------------------\n",
      "Mean number of Reads over Remaining Fusions: 23\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Number of fusions reported in all 3 programs (LongGF, JAFFAL, FusionSeeker): 1\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Groups reported in only 2 programs:\n",
      "Number of fusions reported in exactly 2 programs (LongGF, JAFFAL, FusionSeeker): 1\n",
      "----------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# look over counts in df: fusion_counts_while_filtering\n",
    "final_summary_df = fusion_counts_while_filtering\n",
    "\n",
    "# looking at 'Unique GFs' -- based on gene names - not variable by breakpoint\n",
    "# Drop duplicates based on 'Fusion Name' and 'Sample Name' while keeping the row with the highest 'Supporting Reads'\n",
    "recurrent_filter_df_UNIQUE = recurrent_filter_df.loc[\n",
    "    recurrent_filter_df.groupby(['Fusion Name', 'Sample Name'])['Supporting Reads'].idxmax()]\n",
    "\n",
    "\n",
    "# Total number of GF entries remaining\n",
    "print(\"Results Summary\")\n",
    "print(\"-\" * 50)\n",
    "print(f\"Total number of GF entries remaining: {recurrent_filter_df_UNIQUE.shape[0]}\")\n",
    "print(\"-\" * 50)\n",
    "\n",
    "\n",
    "\n",
    "# Number of Entries without any GFs remaining:\n",
    "# Define the dynamic column name based on percentage_samples\n",
    "recurrent_col_name = f'Filter on Recurrent in {percentage_samples*100}% of Samples'\n",
    "\n",
    "# Number of Entries without any GFs remaining:\n",
    "no_entries_remaining = final_summary_df[final_summary_df[recurrent_col_name] == 0]\n",
    "print(f\"Number of samples with no fusions: {no_entries_remaining.shape[0]}\", no_entries_remaining['Sample Name'].unique())\n",
    "\n",
    "# Number of Entries with exactly ONE GF remaining:\n",
    "one_entry_remaining = final_summary_df[final_summary_df[recurrent_col_name] == 1]\n",
    "print(f\"Number of samples with only one fusion: {one_entry_remaining.shape[0]}\", one_entry_remaining['Sample Name'].unique())\n",
    "\n",
    "# Number of Entries with MORE THAN ONE GF remaining:\n",
    "more_than_one_entry_remaining = final_summary_df[final_summary_df[recurrent_col_name] > 1]\n",
    "print(f\"Number of samples with more than one fusion: {more_than_one_entry_remaining.shape[0]}\", more_than_one_entry_remaining['Sample Name'].unique())\n",
    "\n",
    "print(\"-\" * 50)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# get min, max, and mean number of fusions across samples\n",
    "min_fusions = int(final_summary_df[recurrent_col_name].min())\n",
    "print(f\"Minimum number of fusions: {min_fusions}\")\n",
    "\n",
    "max_fusions = int(final_summary_df[recurrent_col_name].max())\n",
    "print(f\"Maximum number of fusions: {max_fusions}\")\n",
    "\n",
    "mean_fusions = int(final_summary_df[recurrent_col_name].mean())\n",
    "print(f\"Mean number of fusions: {mean_fusions}\")\n",
    "print(\"-\" * 50)\n",
    "\n",
    "# Count of GF Program entries\n",
    "gf_program_counts = recurrent_filter_df_UNIQUE['GF Program'].value_counts()\n",
    "print(\"GF Program Counts:\")\n",
    "print(gf_program_counts)\n",
    "print(\"-\" * 50)\n",
    "\n",
    "# Average number of supporting reads over remaining fusions:\n",
    "mean_fusions_remaining = int(recurrent_filter_df_UNIQUE['Supporting Reads'].mean())\n",
    "print(f\"Mean number of Reads over Remaining Fusions: {mean_fusions_remaining}\")\n",
    "print(\"-\" * 100)\n",
    "\n",
    "\n",
    "#######################################################################################\n",
    "# Looking over the NON-unique entries -- to see if reported in multiple GF Programs \n",
    "# may have varying breakpoint locations \n",
    "#######################################################################################\n",
    "\n",
    "# Overlap in 3 programs (LongGF, JAFFAL, FusionSeeker)\n",
    "gf_programs = ['LongGF', 'JAFFAL', 'FusionSeeker']\n",
    "filtered_df = recurrent_filter_df[recurrent_filter_df['GF Program'].isin(gf_programs)]\n",
    "\n",
    "# Group by fusion, gene symbols, and sample, and count unique GF Programs\n",
    "program_count = filtered_df.groupby(['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol', 'Sample Name'])['GF Program'].nunique()\n",
    "\n",
    "# Filter only those fusions detected in all three programs\n",
    "three_program_count = program_count[program_count == 3].reset_index()\n",
    "\n",
    "# Merge with original DataFrame to get supporting read count\n",
    "three_program_count = three_program_count.merge(\n",
    "    recurrent_filter_df[['Fusion Name', 'Sample Name', 'Supporting Reads']], \n",
    "    on=['Fusion Name', 'Sample Name'], \n",
    "    how='left')\n",
    "\n",
    "# Remove duplicates caused by merging and get max supporting reads per fusion\n",
    "three_program_count = (\n",
    "    three_program_count.groupby(['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol', 'Sample Name'])\n",
    "    .agg({'GF Program': 'first', 'Supporting Reads': 'max'})  # Take max supporting reads\n",
    "    .reset_index())\n",
    "\n",
    "# Sort by Supporting Reads in descending order\n",
    "three_program_count.sort_values(by='Supporting Reads', ascending=False, inplace=True)\n",
    "\n",
    "# Display results\n",
    "print(f\"Number of fusions reported in all 3 programs (LongGF, JAFFAL, FusionSeeker): {three_program_count.shape[0]}\")\n",
    "print(\"-\" * 100)\n",
    "# display(three_program_count)\n",
    "\n",
    "\n",
    "#######################################################################################\n",
    "\n",
    "# Overlap in only 2 programs (LongGF, JAFFAL, FusionSeeker)\n",
    "print(\"Groups reported in only 2 programs:\")\n",
    "\n",
    "# Filter fusions detected in exactly 2 programs\n",
    "two_program_count = program_count[program_count == 2].reset_index()\n",
    "\n",
    "# Merge to get Supporting Reads from original DataFrame\n",
    "two_program_count = two_program_count.merge(\n",
    "    recurrent_filter_df[['Fusion Name', 'Sample Name', 'Supporting Reads']], \n",
    "    on=['Fusion Name', 'Sample Name'], \n",
    "    how='left')\n",
    "\n",
    "# Remove duplicates and get max Supporting Reads per fusion\n",
    "two_program_count = (\n",
    "    two_program_count.groupby(['Fusion Name', 'Gene 1 Symbol', 'Gene 2 Symbol', 'Sample Name'])\n",
    "    .agg({'GF Program': 'first', 'Supporting Reads': 'max'})  # Keep max supporting reads\n",
    "    .reset_index())\n",
    "\n",
    "# Sort by Supporting Reads in descending order\n",
    "two_program_count.sort_values(by='Supporting Reads', ascending=False, inplace=True)\n",
    "\n",
    "# Print summary and display table\n",
    "print(f\"Number of fusions reported in exactly 2 programs (LongGF, JAFFAL, FusionSeeker): {two_program_count.shape[0]}\")\n",
    "print(\"-\" * 100)\n",
    "# display(two_program_count)\n",
    "\n",
    "#######################################################################################\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79eacb3b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1c7b8f4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "fusion_rank",
   "language": "python",
   "name": "fusion_rank"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
