
# Overview of Example Datasets:
This tutorial provides a walkthrough of the `example_chop_fusion_analysis` folder, demonstrating the downstream filtering/analysis of fusion calls from Sample #3 of Cohort 1 sequencing set using the CHOP Cancer Fusion Panel pipeline.


## Example Datasets
**1. CHOP Cancer Fusion Panel Pipeline (Sample 3 – Cohort 1):**  
Located in the `example_chop_fusion_analysis/` folder. This includes:
- Raw fusion calls from each detection tool.
- Combined fusion summaries.
- Filtered fusion outputs using the CHOP Cancer Panel.

**2. Long-Read Whole Transcriptome Fusion Pipeline (Control Sample – Cohort 2):**  
Located in the `example_transcriptome_fusions/` folder. This includes:
- Raw fusion calls from each detection tool.
- Combined fusion summaries.
- Filtered fusion outputs using the transcriptome-specific filtering pipeline.




## Directory Structure

Three folders store program-specific raw fusion call outputs from Sample #3, generated by FusionSeeker, JAFFAL, and LongGF respectively.
- `sample3_fusionseeker/`  
- `sample3_jaffal/`  
- `sample3_longgf/`  

Output files from all three tools after processing through the `01_Combine_Fusion_Calls.ipynb` notebook.
- `combined_fusion_calls/`  

Post-filtering results after applying the CHOP Cancer Fusion Panel filters via `02_CHOP_Cancer_Panel_Filtering.ipynb`.
- `filtered_fusion_calls/`  




---



## Step 1: Combine Program-Specific Fusion Calls

The first step in the workflow uses `01_Combine_Fusion_Calls.ipynb` to summarize the fusion calls across all three tools and generates:

- Individual summaries for each tool:
  - `summary_fusionseeker.tsv`
  - `summary_jaffal.tsv`
  - `summary_longgf.tsv`

- A comprehensive summary across all tools. This summary retains unique gene fusions based on gene names and breakpoint positions.  If multiple tools report the same fusion, the entry with the highest number of supporting reads is retained.
  - `summary_all_fusions.tsv`

---

## Step 2: Apply the CHOP Cancer Panel Filtering Pipeline

`02_CHOP_Cancer_Panel_Filtering.ipynb` filters the combined fusion calls. 

- **Inputs Required:**
  - Fusion call summary file: `summary_all_fusions.tsv`
  - Reference files: found in the `references/` folder
  - Read Support Threshold and Recurrent Rate Percentage
  - Output directory: set to `filtered_fusion_calls/`

- **Output Files Generated:**
  - `summary_filtered_output.tsv`  
    Contains fusion calls that pass through filtering and are annotated using online fusion databases.
  
  - `summary_filtered_output_counts.tsv`  
    Provides a count of how many fusion events remain after each filtering step.

---

## Additional Dataset: Transcriptome Fusion Filtering

A similar analysis was also conducted for a control K562 sample from Cohort 2 sequencing set. This dataset can be found in the `example_transcriptome_fusions/` folder, following a similar approach using the `02_Long_Read_Transcriptome_Filtering.ipynb`. 


